---
title: "Mangrove Bacterial Microbiome"
author: "Golam Rabbani"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Naming convention
## large.variable.A-B.names
## regular_variable_names
## CONSTANT_NAMES
## funtionNames
## ClassNames

knitr::opts_knit$set(root.dir = normalizePath("./Working Files/")) 
```

```{r}
getwd()
```
Loading packages...

```{r}
# Initialising
library(vegan); packageVersion("vegan")
library(phyloseq); packageVersion("phyloseq")
library(dplyr); packageVersion("dplyr")
library(tidyverse); packageVersion("tidyverse")
library(ggplot2); packageVersion("ggplot2")
library(viridis); packageVersion("viridis")
library(svglite); packageVersion("svglite")
library(ggstatsplot); packageVersion("ggstatsplot")
library(ggpubr); packageVersion("ggpubr")
library(scales); packageVersion("scales")
library(RColorBrewer); packageVersion("RColorBrewer")
library(MicEco); packageVersion("MicEco") # For Venn Diagram
library(BiodiversityR); packageVersion("BiodiversityR") # For accumulation curve
library(ggh4x)
library(microbiome)
library(ggnewscale)


getwd()
theme_set(theme_bw())
set.seed(123)



# Color palettes
{
COLOR_21 = c("#e6194b", "#3cb44b", "#ffe119", "#4363d8", "#f58231", 
             "#911eb4", "#46f0f0", "#f032e6", "#bcf60c", "#fabebe", 
             "#008080", "#e6beff", "#9a6324", "#fffac8", "#800000", 
             "#aaffc3", "#808000", "#ffd8b1", "#000075", "#997070", 
             "#000000")
COLOR_30 = c("#fcff5d", "#7dfc00", "#0ec434", "#228c68", "#8ad8e8", 
             "#235b54", "#29bdab", "#3998f5", "#37294f", "#277da7", 
             "#3750db", "#f22020", "#991919", "#ffcba5", "#e68f66", 
             "#c56133", "#96341c", "#632819", "#ffc413", "#f47a22", 
             "#2f2aa0", "#b732cc", "#772b9d", "#f07cab", "#d30b94", 
             "#000000", "#c3a5b4", "#946aa2", "#5d4c86", "#aaaaaa")
COLOR_10 = c("#6b5456", "#ec8d1b", "#6abf2a", "#8b53b7", "#70acbe", 
             "#01c95b", "#c00014", "#31332f", "#f7d000", "#abba00")

QUALITATIVE_COLOR_PAL = brewer.pal.info[brewer.pal.info$category == 'qual',]
COLOR_74 = unlist(mapply(brewer.pal, QUALITATIVE_COLOR_PAL$maxcolors, rownames(QUALITATIVE_COLOR_PAL)))

COLOR_SITE = c("#ec8d1b", "#8b53b7", "#70acbe", "#01c95b", "#c00014", "#f7d000")
NAMES_SITE = c("Kavieng", "Kimbe Bay", "Madang", "Milne Bay", "Port Moresby", "Rabaul")
names(COLOR_SITE) = NAMES_SITE

COLOR_HOST = c("#e6194b", "#3cb44b", "#ffe119", "#4363d8", 
               "#46f0f0", "#f032e6", "#f58231", "#911eb4")
NAMES_HOST = c("Pachyseris speciosa", "Porites lutea", "Diploastrea heliopora", "Pocillopora acuta",
               "Enhalus acoroides", "Thalassia hemprichii", "Avicinea alba", "Sonneratia alba")
names(COLOR_HOST) = NAMES_HOST

COLOR_HOST_TAXA = c("#aaffc3", "#e6beff", "#ffd8b1")
NAMES_HOST_TAXA = c("Mangrove", "Seagrass", "Coral")
names(COLOR_HOST_TAXA) = NAMES_HOST_TAXA

COLOR_HOST_SECTION = c("#9a6324", "#f22020", "#7dfc00", "#228c68", "#ffc413", "#277da7", "#31332f")
NAMES_HOST_SECTION = c("Tissue", "Fruit", "Leaf", "Pneumatophore", "Sediment", "Rhizome", "Root")
names(COLOR_HOST_SECTION) = NAMES_HOST_SECTION
}
```


# Loading phyloseq objects
```{r}
# Loading data
ps.png.bac <- readRDS("./Outputs/PNG_16s/phyloseq_updated-PNG-16s.RDS")
ps.png.bac

## Normalize (relative abundance)
psra.png.bac <- transform_sample_counts(ps.png.bac, function(otu){otu/sum(otu)})

## Filter by taxa group
ps.png.bac.mangrove <- subset_samples(ps.png.bac, hostTaxa == "Mangrove")
ps.png.bac.mangrove.pruned <- prune_taxa(taxa_sums(ps.png.bac.mangrove) > 0, ps.png.bac.mangrove)
psra.png.bac.mangrove <- transform_sample_counts(ps.png.bac.mangrove, function(otu){otu/sum(otu)})
ps.png.bac.mangrove

## Filter by species
ps.png.bac.aa <- subset_samples(ps.png.bac, host == "Avicinea alba")
# ps.png.bac.aa <- prune_taxa(taxa_sums(ps.png.bac.aa) > 0, ps.png.bac.aa)
psra.png.bac.aa <- transform_sample_counts(ps.png.bac.aa, function(otu){otu/sum(otu)})
ps.png.bac.aa

ps.png.bac.sa <- subset_samples(ps.png.bac, host == "Sonneratia alba")
# ps.png.bac.sa <- prune_taxa(taxa_sums(ps.png.bac.sa) > 0, ps.png.bac.sa)
psra.png.bac.sa <- transform_sample_counts(ps.png.bac.sa, function(otu){otu/sum(otu)})
ps.png.bac.sa

ps.png.bac.mangrove.fruit <- subset_samples(ps.png.bac.mangrove, hostSection == "Fruit")
psra.png.bac.mangrove.fruit <- transform_sample_counts(ps.png.bac.mangrove.fruit, function(otu){otu/sum(otu)})
ps.png.bac.mangrove.fruit

ps.png.bac.mangrove.leaf <- subset_samples(ps.png.bac.mangrove, hostSection == "Leaf")
psra.png.bac.mangrove.leaf <- transform_sample_counts(ps.png.bac.mangrove.leaf, function(otu){otu/sum(otu)})
ps.png.bac.mangrove.leaf

ps.png.bac.mangrove.pneumatophore <- subset_samples(ps.png.bac.mangrove, hostSection == "Pneumatophore")
psra.png.bac.mangrove.pneumatophore <- transform_sample_counts(ps.png.bac.mangrove.pneumatophore, function(otu){otu/sum(otu)})
ps.png.bac.mangrove.pneumatophore

ps.png.bac.mangrove.sediment <- subset_samples(ps.png.bac.mangrove, hostSection == "Sediment")
psra.png.bac.mangrove.sediment <- transform_sample_counts(ps.png.bac.mangrove.sediment, function(otu){otu/sum(otu)})
ps.png.bac.mangrove.sediment

ps.png.bac.mangrove.nosediment <- subset_samples(ps.png.bac.mangrove, hostSection != "Sediment")
psra.png.bac.mangrove.nosediment <- transform_sample_counts(ps.png.bac.mangrove.nosediment, function(otu){otu/sum(otu)})
ps.png.bac.mangrove.nosediment
```

```{r}
otu.mangrove <- as.data.frame(ps.png.bac.mangrove@otu_table)

meta.mangrove <- as.data.frame(ps.png.bac.mangrove@sam_data)
```

# Rarefaction Curves
```{r}
## Functions
rarefactionCurve = function(ps, group, color_pal, filename_tag = "TAG"){
  otu_table_matrix <- ps@otu_table
  class(otu_table_matrix) <- "matrix"
  rarefaction_df <- rarecurve(otu_table_matrix, step = 20, label = FALSE, tidy = TRUE)
  rarefaction_df$group <- sample_data(ps)[rarefaction_df$Site,][[group]]
  rarefaction_df <- rarefaction_df %>% group_by(Site)
  rarefaction_curve <- ggplot(rarefaction_df) + geom_line(aes(x = Sample, y = Species, group = Site, color = group), size = 0.5, alpha = 0.5)
  ggsave(rarefaction_curve, filename = paste0("./Outputs/rarefactionCurves-", group, "-", filename_tag, ".svg"), dpi = 300, width = 12, height = 8)
  rarefaction_curve
}
```

```{r}
rarefactionCurve(ps.png.bac.mangrove, "site", COLOR_SITE, filename_tag = "PNG-Mangrove-16s")
rarefactionCurve(ps.png.bac.mangrove, "host", COLOR_HOST, filename_tag = "PNG-Mangrove-16s")
```

# Preston model for estimated diversity coverage

```{r}
preston.mod.oct <- prestonfit(colSums(otu.mangrove))
preston.mod.ll <- prestondistr(colSums(otu.mangrove))
preston.mod.oct
preston.mod.ll

plot(preston.mod.oct)  
lines(preston.mod.ll, line.col = "blue3") # Different
## Smoothed density
den <- density(log2(colSums(otu.mangrove)))
lines(den$x, ncol(otu.mangrove)*den$y, lwd=2) # Fairly similar to mod.oct

## Extrapolated richness
veiledspec(preston.mod.oct)
veiledspec(preston.mod.oct)[2] / veiledspec(preston.mod.oct)[1]
veiledspec(preston.mod.ll)
veiledspec(preston.mod.ll)[2] / veiledspec(preston.mod.ll)[1]
```

# Accumulation curve of richness
```{r}
venn_mangrove_hostSection <- ps_venn(ps.png.bac.mangrove, "hostSection")#, fill = alpha(COLOR_HOST_SECTION, 0.5))
venn_mangrove_hostSection
```

```{r}
accum_curve_exact = specaccum(otu.mangrove, method = "exact")

table(meta.mangrove$site[order(meta.mangrove$site)])
accum_curve_site = specaccum(otu.mangrove[order(meta.mangrove$site),], method = "collector")
ggplot() + 
  annotate("rect", xmin = 0, xmax = 40.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_SITE["Kavieng"]) +
  annotate("rect", xmin = 40.5, xmax = 120.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_SITE["Kimbe Bay"]) +
  annotate("rect", xmin = 120.5, xmax = 200.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_SITE["Madang"]) +
  annotate("rect", xmin = 200.5, xmax = 320.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_SITE["Milne Bay"]) +
  annotate("rect", xmin = 320.5, xmax = 400, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_SITE["Port Moresby"]) +
  geom_line(aes(x = accum_curve_site$sites, y = accum_curve_site$richness)) +
  lims(x = c(0,400))
  

unique(meta.mangrove$host[order(meta.mangrove$host)])
accum_curve_host = specaccum(otu.mangrove[order(meta.mangrove$host),], method = "collector")


table(meta.mangrove$hostSection[order(meta.mangrove$hostSection)])
accum_curve_hostSection <- specaccum(otu.mangrove[order(meta.mangrove$hostSection),], method = "collector")
ggplot() + 
  annotate("rect", xmin = 0, xmax = 100.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Fruit"]) +
  annotate("rect", xmin = 100.5, xmax = 200.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Leaf"]) +
  annotate("rect", xmin = 200.5, xmax = 300.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Pneumatophore"]) +
  annotate("rect", xmin = 300.5, xmax = 400, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Sediment"]) +
  geom_line(aes(x = accum_curve_hostSection$sites, y = accum_curve_hostSection$richness)) +
  lims(x = c(0,400))


accum_curve_hostSection_exact_fruit <- specaccum(otu.mangrove, subset = meta.mangrove$hostSection == "Fruit", method = "exact")
accum_curve_hostSection_exact_leaf <- specaccum(otu.mangrove, subset = meta.mangrove$hostSection == "Leaf", method = "exact")
accum_curve_hostSection_exact_pneumatophore <- specaccum(otu.mangrove, subset = meta.mangrove$hostSection == "Pneumatophore", method = "exact")
accum_curve_hostSection_exact_sediment <- specaccum(otu.mangrove, subset = meta.mangrove$hostSection == "Sediment", method = "exact")

df <- data.frame(sites = 
                   c(accum_curve_hostSection_exact_sediment$sites,
                     accum_curve_hostSection_exact_pneumatophore$sites + 
                       max(),
                     accum_curve_hostSection_exact_leaf$sites + 
                       max(),
                     accum_curve_hostSection_exact_fruit$sites + 
                       max()
                     ),
                 richness = 
                   c(accum_curve_hostSection_exact_sediment$richness,
                     accum_curve_hostSection_exact_pneumatophore$richness + 
                       max(),
                     accum_curve_hostSection_exact_leaf$richness + 
                       max(),
                     accum_curve_hostSection_exact_fruit$richness + 
                       max()
                     )
                 )
ggplot() + 
  annotate("rect", xmin = 0, xmax = 100.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Fruit"]) +
  annotate("rect", xmin = 100.5, xmax = 200.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Leaf"]) +
  annotate("rect", xmin = 200.5, xmax = 300.5, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Pneumatophore"]) +
  annotate("rect", xmin = 300.5, xmax = 400, ymin = -Inf, ymax = Inf,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Sediment"]) +
  geom_line(aes(x = df$sites, y = df$richness)) +
  lims(x = c(0,400))

unique(as.data.frame(ps.png.bac@sam_data)$hostTaxa[order(as.data.frame(ps.png.bac@sam_data)$hostTaxa)])
accum_curve_hostTaxa = specaccum(as.data.frame(ps.png.bac@otu_table)[order(as.data.frame(ps.png.bac@sam_data)$site),], method = "collector")
```

```{r}

```

# Top 20 ASVs Filtering
```{r}
## Function
createTop20ASVPhyloseq <- function(ps){
  taxa_abun <- as.data.frame(colSums(ps@otu_table))
  taxa_abun <- taxa_abun %>% 
    mutate(
      ASV = rownames(taxa_abun),
      `colSums(ps@otu_table)` = NULL
    )
  sample_otu <- as.data.frame(ps@otu_table)
  sum_otu <- aggregate(sample_otu, by = list(ps@sam_data$host), sum)
  rownames(sum_otu) <- sum_otu[,1]
  sum_otu <- as.data.frame(t(sum_otu[,-1]))
  top_20 <- c()
  for(i in unique(ps@sam_data$host)){
    top_20 <- append(top_20, rownames(slice_max(sum_otu, get(i), n = 20)))
  }
  top_20 <- unique(top_20)
  ps_not_top_20 <- prune_taxa(!(taxa_names(ps) %in% top_20), ps)
  other_abun <- rowSums(ps_not_top_20@otu_table)
  ps_top_20 <- prune_taxa(taxa_names(ps) %in% top_20, ps)
  otu_top_20 <- as.data.frame(otu_table(ps_top_20, taxa_are_rows = FALSE))
  identical(rownames(otu_top_20), names(other_abun))
  otu_top_20$ASVOthers <- other_abun
  taxa_top_20 <- as.data.frame(tax_table(ps_top_20))
  taxa_top_20 <- mutate(taxa_top_20, across(everything(), ~replace_na(.x, "Unidentified")))
  ASVOthers <- rep("Other", 7)
  taxa_top_20[length(top_20) + 1,] <- ASVOthers
  rownames(taxa_top_20)[length(top_20) + 1] <- "ASVOthers"
  taxa_top_20 <- as.matrix(taxa_top_20)
  ps_top_20 <- phyloseq(otu_table(otu_top_20, taxa_are_rows = FALSE), 
                                  sample_data(ps_top_20),
                                  tax_table(taxa_top_20))
  return(ps_top_20)
}
```

```{r}
## All
ps.png.bac.top20 <- createTop20ASVPhyloseq(ps.png.bac)

## Mangrove
ps.png.bac.mangrove.top20 <- createTop20ASVPhyloseq(ps.png.bac.mangrove)

## By Species
ps.png.bac.aa.top20 <- createTop20ASVPhyloseq(ps.png.bac.aa)
ps.png.bac.sa.top20 <- createTop20ASVPhyloseq(ps.png.bac.sa)

## Mapping for colors
taxa_top_20_all <- as.data.frame(ps.png.bac.top20@tax_table)
COLOR_PHYLUM <- setNames(c(COLOR_21[1:(length(unique(taxa_top_20_all$Phylum)) - 2)], "#a1a1a1", "#e1e1e1"), c(sort(setdiff(unique(taxa_top_20_all$Phylum), c("Other", "Unidentified"))), "Unidentified", "Other"))
COLOR_CLASS <- setNames(c(COLOR_21[1:(length(unique(taxa_top_20_all$Class)) - 2)], "#a1a1a1","#e1e1e1"), c(sort(setdiff(unique(taxa_top_20_all$Class), c("Other", "Unidentified"))), "Unidentified", "Other"))
COLOR_ORDER <- setNames(c(COLOR_30[1:(length(unique(taxa_top_20_all$Order)) - 2)], "#a1a1a1","#e1e1e1"), c(sort(setdiff(unique(taxa_top_20_all$Order), c("Other", "Unidentified"))), "Unidentified", "Other"))
COLOR_FAMILY <- setNames(c(COLOR_74[1:(length(unique(taxa_top_20_all$Family)) - 2)], "#a1a1a1","#e1e1e1"), c(sort(setdiff(unique(taxa_top_20_all$Family), c("Other", "Unidentified"))), "Unidentified", "Other"))
COLOR_GENUS <-  setNames(c(COLOR_74[1:(length(unique(taxa_top_20_all$Genus)) - 2)], "#a1a1a1","#e1e1e1"), c(sort(setdiff(unique(taxa_top_20_all$Genus), c("Other", "Unidentified"))), "Unidentified", "Other"))
COLOR_SPECIES <-  setNames(c(COLOR_10[1:(length(unique(taxa_top_20_all$Species)) - 2)], "#a1a1a1","#e1e1e1"), c(sort(setdiff(unique(taxa_top_20_all$Species), c("Other", "Unidentified"))), "Unidentified", "Other"))
```

# Relative Abundance
```{r}
## Functions
simpleBarplot = function(ps, x = "Sample", y = "Abundance", fill = NULL, color_pal = NULL, title = NULL, facet_wrap = NULL) {
  # Defining function to make bar charts without black lines separating samples. Based on phyloseq function "plot_bar".
  mdf = psmelt(ps)
  p = ggplot(mdf, aes_string(x = x, y = y, fill = if(TRUE){factor(mdf[,fill], levels = names(color_pal))}else{fill}))
  p = p + geom_bar(stat = "identity", size = 0)#, position = "stack")
  p = p + theme(axis.text = element_text(size = 15), 
                axis.title = element_text(size = 17,face = "bold"), 
                axis.text.x = element_text(angle = 70, hjust = 1))
  p = p + labs(y = "Relative Abundance", fill = fill)
  p = p + guides(guide_legend(ncol = 1), fill = guide_legend(ncol = 3))
  p = p + scale_y_continuous(expand = c(0,0), n.breaks = 6)
  if (!is.null(facet_wrap)) {
    p <- p + facet_wrap(facet_wrap, nrow = 1, scales = "free_x") + theme(strip.text = element_text(size = 15))
  }
  if (!is.null(title)) {
    p <- p + ggtitle(title)
  }
  return(p)
}

makeGrouppsra <- function(ps.variable, group.variable){
  combined_df <- as.data.frame(ps.variable@otu_table)
  combined_df[[group.variable]] <- ps.variable@sam_data[[group.variable]]
  group_otu <- combined_df[,-ncol(combined_df)] %>%
    group_by(combined_df[[group.variable]]) %>%
    summarise_all(.funs = sum)
  group_otu <- as.data.frame(group_otu)
  row.names(group_otu) <- paste0("Samples_", group_otu[[1]])
  group_meta <- data.frame(group.variable = group_otu[[1]])
  colnames(group_meta) <- group.variable
  row.names(group_meta) <- NULL
  row.names(group_meta) <- paste0("Samples_", group_otu[[1]])
  group_otu <- group_otu[,-1]
  group_ps <- phyloseq(otu_table(group_otu, taxa_are_rows = FALSE), 
                               sample_data(group_meta), 
                               tax_table(ps.variable@tax_table))
  group_psra <- transform_sample_counts(group_ps, function(otu){otu/sum(otu)})
  return(group_psra)
}

plotRelativeAbundanceBarplot <- function(ps.variable, x.variable = "sampleName", fill.variable = "Phylum", color_pal = COLOR_30, title.variable = "General"){
  if(x.variable == "sampleName"){ 
    psra <- transform_sample_counts(ps.variable, function(otu){otu/sum(otu)})
  }else{
    psra <- makeGrouppsra(ps.variable, x.variable)
  }
  ra_barplot <- simpleBarplot(psra, x = x.variable, fill = fill.variable, color_pal = if(!identical(color_pal, COLOR_30)){color_pal}else{NULL}) + 
    theme(legend.text = element_text(size = rel(1.5)), 
          legend.title = element_text(size = rel(1.5)), 
          legend.key.size = unit(0.3, "line")) + 
    guides(fill = guide_legend(nrow = min(c(80, length(unique(as.data.frame(psra@tax_table)[[fill.variable]])))))) + 
    scale_fill_manual(values = color_pal) +
    ggtitle(paste(title.variable, fill.variable, "Relative Abundance"))
  return(ra_barplot)
}
```

## Phylum
```{r}
### Mangrove
ra_barplot_bac_mangrove_host_phylum <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "host", fill.variable = "Phylum", color_pal = COLOR_PHYLUM, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_host_phylum, filename = "./Outputs/relativeAbundanceBarplot-magrove-host-phylum-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_host_phylum
ra_barplot_bac_mangrove_host_phylum_full <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove, x.variable = "host", fill.variable = "Phylum", title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_host_phylum_full, filename = "./Outputs/relativeAbundanceBarplot-magrove-host-phylum-PNG-16s-full.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_host_phylum_full
ra_barplot_bac_mangrove_site_phylum <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "site", fill.variable = "Phylum", color_pal = COLOR_PHYLUM, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_site_phylum, filename = "./Outputs/relativeAbundanceBarplot-mangrove-site-phylum-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_site_phylum
ra_barplot_bac_mangrove_sample_phylum <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "sampleName", fill.variable = "Phylum", color_pal = COLOR_PHYLUM, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_sample_phylum, filename = "./Outputs/relativeAbundanceBarplot-mangrove-sample-phylum-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_sample_phylum


  
### By Species
#### Avicinea alba
ra_barplot_bac_aa_site_phylum <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "site", fill.variable = "Phylum", color_pal = COLOR_PHYLUM, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_site_phylum, filename = "./Outputs/relativeAbundanceBarplot-aa-site-phylum-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_site_phylum
ra_barplot_bac_aa_sample_phylum <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "sampleName", fill.variable = "Phylum", color_pal = COLOR_PHYLUM, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_sample_phylum, filename = "./Outputs/relativeAbundanceBarplot-aa-sample-phylum-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_sample_phylum
  
#### Sonneratia alba
ra_barplot_bac_sa_site_phylum <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "site", fill.variable = "Phylum", color_pal = COLOR_PHYLUM, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_site_phylum, filename = "./Outputs/relativeAbundanceBarplot-sa-site-phylum-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_site_phylum
ra_barplot_bac_sa_sample_phylum <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "sampleName", fill.variable = "Phylum", color_pal = COLOR_PHYLUM, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_sample_phylum, filename = "./Outputs/relativeAbundanceBarplot-sa-sample-phylum-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_sample_phylum

# Obtaining top 20 phyla
ps.png.bac.mangrove.phylum <- tax_glom(ps.png.bac.mangrove, taxrank = "Phylum", NArm = FALSE)
ps.png.bac.mangrove.phylum
ps.png.bac.mangrove.phylum.top20 <- createTop20ASVPhyloseq(ps.png.bac.mangrove.phylum)

phylum_top_20 <- as.data.frame(ps.png.bac.mangrove.phylum.top20@tax_table)
COLOR_PHYLUM_TOP20 <- setNames(c(COLOR_30[1:(length(unique(phylum_top_20$Phylum)) - 2)], "#a1a1a1", "#e1e1e1"), c(sort(setdiff(unique(phylum_top_20$Phylum), c("Other", "Unidentified"))), "Unidentified", "Other"))

ra_barplot_bac_mangrove_phylum <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.phylum.top20, x.variable = "sampleName", fill.variable = "Phylum", color_pal = COLOR_PHYLUM_TOP20, title.variable = "Mangrove") + facet_grid2(hostSection~site+host, scales="free", independent="y") + coord_flip() + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank()) + theme(legend.position = "bottom") + guides(fill = guide_legend(ncol = 5))
ggsave(ra_barplot_bac_mangrove_phylum, filename = "./Outputs/relativeAbundanceBarplot-mangrove-phylum-PNG-16s.svg", dpi = 300, width = 16, height = 10)
ra_barplot_bac_mangrove_phylum

# Calculating dominance of phyla
ps.png.bac.mangrove.phylum
phylum_abundance <- psmelt(ps.png.bac.mangrove.phylum)

phylum_summary <- phylum_abundance %>%
  group_by(Phylum) %>%
  summarise(Total = sum(Abundance)) %>%
  mutate(RelativeAbundance = Total / sum(Total))
phylum_summary %>%
  arrange(desc(RelativeAbundance))

phylum_abundance_fruit <- phylum_abundance[phylum_abundance$hostSection == "Fruit",]
phylum_summary <- phylum_abundance_fruit %>%
  group_by(Phylum) %>%
  summarise(Total = sum(Abundance)) %>%
  mutate(RelativeAbundance = Total / sum(Total))
phylum_summary %>%
  arrange(desc(RelativeAbundance))

phylum_abundance_leaf <- phylum_abundance[phylum_abundance$hostSection == "Leaf",]
phylum_summary <- phylum_abundance_leaf %>%
  group_by(Phylum) %>%
  summarise(Total = sum(Abundance)) %>%
  mutate(RelativeAbundance = Total / sum(Total))
phylum_summary %>%
  arrange(desc(RelativeAbundance))

phylum_abundance_pneumatophore <- phylum_abundance[phylum_abundance$hostSection == "Pneumatophore",]
phylum_summary <- phylum_abundance_pneumatophore %>%
  group_by(Phylum) %>%
  summarise(Total = sum(Abundance)) %>%
  mutate(RelativeAbundance = Total / sum(Total))
phylum_summary %>%
  arrange(desc(RelativeAbundance))

phylum_abundance_sediment <- phylum_abundance[phylum_abundance$hostSection == "Sediment",]
phylum_summary <- phylum_abundance_sediment %>%
  group_by(Phylum) %>%
  summarise(Total = sum(Abundance)) %>%
  mutate(RelativeAbundance = Total / sum(Total))
phylum_summary %>%
  arrange(desc(RelativeAbundance))
```

## Class
```{r}
### Mangrove
ra_barplot_bac_mangrove_host_class <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "host", fill.variable = "Class", color_pal = COLOR_CLASS, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_host_class, filename = "./Outputs/relativeAbundanceBarplot-magrove-host-class-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_host_class
ra_barplot_bac_mangrove_site_class <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "site", fill.variable = "Class", color_pal = COLOR_CLASS, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_site_class, filename = "./Outputs/relativeAbundanceBarplot-mangrove-site-class-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_site_class
ra_barplot_bac_mangrove_sample_class <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "sampleName", fill.variable = "Class", color_pal = COLOR_CLASS, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_sample_class, filename = "./Outputs/relativeAbundanceBarplot-mangrove-sample-class-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_sample_class

### By Species
#### Avicinea alba
ra_barplot_bac_aa_site_class <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "site", fill.variable = "Class", color_pal = COLOR_CLASS, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_site_class, filename = "./Outputs/relativeAbundanceBarplot-aa-site-class-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_site_class
ra_barplot_bac_aa_sample_class <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "sampleName", fill.variable = "Class", color_pal = COLOR_CLASS, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_sample_class, filename = "./Outputs/relativeAbundanceBarplot-aa-sample-class-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_sample_class
  
#### Sonneratia alba
ra_barplot_bac_sa_site_class <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "site", fill.variable = "Class", color_pal = COLOR_CLASS, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_site_class, filename = "./Outputs/relativeAbundanceBarplot-sa-site-class-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_site_class
ra_barplot_bac_sa_sample_class <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "sampleName", fill.variable = "Class", color_pal = COLOR_CLASS, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_sample_class, filename = "./Outputs/relativeAbundanceBarplot-sa-sample-class-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_sample_class
```

## Order
```{r}
### Mangrove
ra_barplot_bac_mangrove_host_order <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "host", fill.variable = "Order", color_pal = COLOR_ORDER, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_host_order, filename = "./Outputs/relativeAbundanceBarplot-magrove-host-order-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_host_order
ra_barplot_bac_mangrove_site_order <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "site", fill.variable = "Order", color_pal = COLOR_ORDER, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_site_order, filename = "./Outputs/relativeAbundanceBarplot-mangrove-site-order-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_site_order
ra_barplot_bac_mangrove_sample_order <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "sampleName", fill.variable = "Order", color_pal = COLOR_ORDER, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_sample_order, filename = "./Outputs/relativeAbundanceBarplot-mangrove-sample-order-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_sample_order

### By Species
#### Avicinea alba
ra_barplot_bac_aa_site_order <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "site", fill.variable = "Order", color_pal = COLOR_ORDER, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_site_order, filename = "./Outputs/relativeAbundanceBarplot-aa-site-order-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_site_order
ra_barplot_bac_aa_sample_order <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "sampleName", fill.variable = "Order", color_pal = COLOR_ORDER, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_sample_order, filename = "./Outputs/relativeAbundanceBarplot-aa-sample-order-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_sample_order
  
#### Sonneratia alba
ra_barplot_bac_sa_site_order <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "site", fill.variable = "Order", color_pal = COLOR_ORDER, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_site_order, filename = "./Outputs/relativeAbundanceBarplot-sa-site-order-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_site_order
ra_barplot_bac_sa_sample_order <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "sampleName", fill.variable = "Order", color_pal = COLOR_ORDER, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_sample_order, filename = "./Outputs/relativeAbundanceBarplot-sa-sample-order-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_sample_order
```

## Family
```{r}
### Mangrove
ra_barplot_bac_mangrove_host_family <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "host", fill.variable = "Family", color_pal = COLOR_FAMILY, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_host_family, filename = "./Outputs/relativeAbundanceBarplot-magrove-host-family-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_host_family
ra_barplot_bac_mangrove_site_family <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "site", fill.variable = "Family", color_pal = COLOR_FAMILY, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_site_family, filename = "./Outputs/relativeAbundanceBarplot-mangrove-site-family-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_site_family
ra_barplot_bac_mangrove_sample_family <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "sampleName", fill.variable = "Family", color_pal = COLOR_FAMILY, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_sample_family, filename = "./Outputs/relativeAbundanceBarplot-mangrove-sample-family-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_sample_family

### By Species
#### Avicinea alba
ra_barplot_bac_aa_site_family <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "site", fill.variable = "Family", color_pal = COLOR_FAMILY, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_site_family, filename = "./Outputs/relativeAbundanceBarplot-aa-site-family-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_site_family
ra_barplot_bac_aa_sample_family <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "sampleName", fill.variable = "Family", color_pal = COLOR_FAMILY, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_sample_family, filename = "./Outputs/relativeAbundanceBarplot-aa-sample-family-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_sample_family
  
#### Sonneratia alba
ra_barplot_bac_sa_site_family <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "site", fill.variable = "Family", color_pal = COLOR_FAMILY, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_site_family, filename = "./Outputs/relativeAbundanceBarplot-sa-site-family-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_site_family
ra_barplot_bac_sa_sample_family <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "sampleName", fill.variable = "Family", color_pal = COLOR_FAMILY, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_sample_family, filename = "./Outputs/relativeAbundanceBarplot-sa-sample-family-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_sample_family
```

## Genus
```{r}
### Mangrove
ra_barplot_bac_mangrove_host_genus <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "host", fill.variable = "Genus", color_pal = COLOR_GENUS, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_host_genus, filename = "./Outputs/relativeAbundanceBarplot-magrove-host-genus-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_host_genus
ra_barplot_bac_mangrove_site_genus <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "site", fill.variable = "Genus", color_pal = COLOR_GENUS, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_site_genus, filename = "./Outputs/relativeAbundanceBarplot-mangrove-site-genus-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_site_genus
ra_barplot_bac_mangrove_sample_genus <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "sampleName", fill.variable = "Genus", color_pal = COLOR_GENUS, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_sample_genus, filename = "./Outputs/relativeAbundanceBarplot-mangrove-sample-genus-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_sample_genus

### By Species
#### Avicinea alba
ra_barplot_bac_aa_site_genus <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "site", fill.variable = "Genus", color_pal = COLOR_GENUS, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_site_genus, filename = "./Outputs/relativeAbundanceBarplot-aa-site-genus-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_site_genus
ra_barplot_bac_aa_sample_genus <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "sampleName", fill.variable = "Genus", color_pal = COLOR_GENUS, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_sample_genus, filename = "./Outputs/relativeAbundanceBarplot-aa-sample-genus-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_sample_genus
  
#### Sonneratia alba
ra_barplot_bac_sa_site_genus <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "site", fill.variable = "Genus", color_pal = COLOR_GENUS, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_site_genus, filename = "./Outputs/relativeAbundanceBarplot-sa-site-genus-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_site_genus
ra_barplot_bac_sa_sample_genus <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "sampleName", fill.variable = "Genus", color_pal = COLOR_GENUS, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_sample_genus, filename = "./Outputs/relativeAbundanceBarplot-sa-sample-genus-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_sample_genus
```

## Species
```{r}
### Mangrove
ra_barplot_bac_mangrove_host_species <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "host", fill.variable = "Species", color_pal = COLOR_SPECIES, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_host_species, filename = "./Outputs/relativeAbundanceBarplot-mangrove-host-species-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_host_species
ra_barplot_bac_mangrove_site_species <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "site", fill.variable = "Species", color_pal = COLOR_SPECIES, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_site_species, filename = "./Outputs/relativeAbundanceBarplot-mangrove-site-species-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_site_species
ra_barplot_bac_mangrove_sample_species <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.mangrove.top20, x.variable = "sampleName", fill.variable = "Species", color_pal = COLOR_SPECIES, title.variable = "Mangrove")
ggsave(ra_barplot_bac_mangrove_sample_species, filename = "./Outputs/relativeAbundanceBarplot-mangrove-sample-species-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_mangrove_sample_species

### By Species
#### Avicinea alba
ra_barplot_bac_aa_site_species <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "site", fill.variable = "Species", color_pal = COLOR_SPECIES, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_site_species, filename = "./Outputs/relativeAbundanceBarplot-aa-site-species-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_site_species
ra_barplot_bac_aa_sample_species <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.aa.top20, x.variable = "sampleName", fill.variable = "Species", color_pal = COLOR_SPECIES, title.variable = "Avicinea alba")
ggsave(ra_barplot_bac_aa_sample_species, filename = "./Outputs/relativeAbundanceBarplot-aa-sample-species-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_aa_sample_species
  
#### Sonneratia alba
ra_barplot_bac_sa_site_species <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "site", fill.variable = "Species", color_pal = COLOR_SPECIES, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_site_species, filename = "./Outputs/relativeAbundanceBarplot-sa-site-species-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_site_species
ra_barplot_bac_sa_sample_species <- plotRelativeAbundanceBarplot(ps.variable = ps.png.bac.sa.top20, x.variable = "sampleName", fill.variable = "Species", color_pal = COLOR_SPECIES, title.variable = "Sonneratia alba")
ggsave(ra_barplot_bac_sa_sample_species, filename = "./Outputs/relativeAbundanceBarplot-sa-sample-species-PNG-16s.svg", dpi = 300, width = 12, height = 10)
ra_barplot_bac_sa_sample_species
```

# Diversity
```{r}
diversity <- data.frame(
                  site = psra.png.bac.mangrove@sam_data$site,
                  host = psra.png.bac.mangrove@sam_data$host,
                  hostTaxa = psra.png.bac.mangrove@sam_data$hostTaxa,
                  hostSection = psra.png.bac.mangrove@sam_data$hostSection,
                  shannon = vegan::diversity(otu_table(psra.png.bac.mangrove, taxa_are_rows = FALSE)),
                  richness = t(colSums(decostand(otu_table(psra.png.bac.mangrove, taxa_are_rows = FALSE), method = "pa"))))
write.csv(diversity, file = "./Outputs/diversity-mangrove-PNG-16s.csv", quote = FALSE)
diversity

## By site
diversity %>% group_by(site) %>%
  summarise(N = n(), Min = min(shannon), Max = max(shannon), Mean = mean(shannon), Median = median(shannon), SD = sd(shannon), )

## By host
diversity %>% group_by(host) %>%
  summarise(N = n(), Min = min(shannon), Max = max(shannon), Mean = mean(shannon), Median = median(shannon), SD = sd(shannon), )

## By hostTaxa
diversity %>% group_by(hostTaxa) %>%
  summarise(N = n(), Min = min(shannon), Max = max(shannon), Mean = mean(shannon), Median = median(shannon), SD = sd(shannon), )

## By hostSection
diversity %>% group_by(hostSection) %>%
  summarise(N = n(), Min = min(shannon), Max = max(shannon), Mean = mean(shannon), Median = median(shannon), SD = sd(shannon), )
```

```{r}
## Boxplots
ggplot(diversity, aes(x = site, y = shannon, fill = host)) + 
  geom_boxplot() + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Site",
    y = "Shannon Diversity",
    fill = "Host"
    ) +
  scale_fill_manual(values = COLOR_HOST)
ggsave(filename = "./Outputs/shannonDiversity-filled-site-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)

ggplot(diversity, aes(x=site, y=shannon, fill = host)) + 
  geom_violin(draw_quantiles = c(0.25, 0.50, 0.75)) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Site",
    y = "Shannon Diversity",
    fill = "Host"
  ) +
  scale_fill_manual(values = COLOR_HOST)

ggplot(diversity, aes(x = host, y = shannon, fill = site)) + 
  geom_boxplot() + theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Host",
    y = "Shannon Diversity",
    fill = "Site"
  ) +
  scale_fill_manual(values = COLOR_SITE)
ggsave(filename = "./Outputs/shannonDiversity-filled-host-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)

ggplot(diversity, aes(x = host, y = shannon, fill = hostTaxa)) + 
  geom_boxplot() + theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Host",
    y = "Shannon Diversity",
    fill = "Host Taxa"
  ) +
  scale_fill_manual(values = COLOR_HOST_TAXA)
ggsave(filename = "./Outputs/shannonDiversity-filled-hostTaxa-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)

ggplot(diversity, aes(x = host, y = shannon, fill = hostSection)) + 
  geom_boxplot() + theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Host",
    y = "Shannon Diversity",
    fill = "Host Section"
  ) +
  scale_fill_manual(values = COLOR_HOST_SECTION)
ggsave(filename = "./Outputs/shannonDiversity-filled-host-hostSection-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)

ggplot(diversity, aes(x = site, y = shannon, fill = hostSection)) + 
  geom_boxplot() + theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Site",
    y = "Shannon Diversity",
    fill = "Host Section"
  ) +
  scale_fill_manual(values = COLOR_HOST_SECTION) +
  facet_wrap(vars(host))
ggsave(filename = "./Outputs/shannonDiversity-filled-site-hostSection-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)
```

```{r}
# New Boxplots

## For Site
ggplot(diversity, aes(x = site, y = shannon)) + 
  geom_jitter(aes(color = site), alpha = 0.5, height = 0, size = 5) +
  geom_boxplot(alpha = 0) + 
  theme_bw() + 
  theme(axis.text.x = element_blank(), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Site",
    y = "Shannon Diversity",
    color = "Site"
    ) +
  scale_color_manual(values = COLOR_SITE)
ggsave(filename = "./Outputs/shannonDiversity-site-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)

### with host as color
ggplot(diversity, aes(x = site, y = shannon)) + 
  geom_jitter(aes(color = host), alpha = 0.5, height = 0, size = 5) +
  geom_boxplot(alpha = 0) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Site",
    y = "Shannon Diversity",
    color = "Host"
    ) +
  scale_color_manual(values = COLOR_HOST)
ggsave(filename = "./Outputs/shannonDiversity-site-host-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)

## For Host
ggplot(diversity, aes(x = host, y = shannon)) + 
  geom_jitter(aes(color = host), alpha = 0.5, height = 0, size = 5) +
  geom_boxplot(alpha = 0) + 
  theme_bw() + 
  theme(axis.text.x = element_blank(), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Host",
    y = "Shannon Diversity",
    color = "Host"
    ) +
  scale_color_manual(values = COLOR_HOST)
ggsave(filename = "./Outputs/shannonDiversity-host-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)

## For HostSection for separate species
ggplot(diversity, aes(x = hostSection, y = shannon)) + 
  geom_jitter(aes(color = site, shape = hostSection), alpha = 0.5, height = 0, size = 5) +
  geom_boxplot(alpha = 0) + 
  theme_bw() + 
  theme(axis.text.x = element_blank(), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Host",
    y = "Shannon Diversity",
    color = "Site",
    shape = "Host Section"
    ) +
  scale_color_manual(values = COLOR_SITE) +
  facet_wrap(vars(host))
ggsave(filename = "./Outputs/shannonDiversity-hostSection-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)

## For site for separate hostsection and species
ggplot(diversity, aes(x = site, y = shannon)) + 
  geom_jitter(aes(color = site, shape = host), alpha = 0.5, height = 0, size = 5) +
  geom_boxplot(alpha = 0) + 
  theme_bw() + 
  theme(axis.text.x = element_blank(), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Host",
    y = "Shannon Diversity",
    color = "Site",
    shape = "Host"
    ) +
  scale_color_manual(values = COLOR_SITE) +
  facet_wrap(vars(hostSection))
ggsave(filename = "./Outputs/shannonDiversity-site-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)

ggplot(diversity, aes(x = site, color = host, y = shannon)) + 
  geom_boxplot(aes(fill = host), alpha = 10) + 
  geom_jitter(aes(color = site, shape = host), alpha = 0.5, height = 0, size = 5) +
  theme_bw() + 
  theme(axis.text.x = element_blank(), axis.text = element_text(size = 25), axis.title.y = element_text(vjust = 2), axis.title = element_text(size = 30), title = element_text(size = 17), legend.text = element_text(size = 25), legend.title = element_text(size = 30)) + 
  labs(
    x = "Host",
    y = "Shannon Diversity",
    color = "Site",
    shape = "Host"
    ) +
  scale_color_manual(values = COLOR_SITE) +
  scale_fill_manual(values = COLOR_HOST) +
  facet_wrap(vars(hostSection))
ggsave(filename = "./Outputs/shannonDiversity-site_sepHost-mangrove-PNG-16s.svg", dpi = 300, width = 12, height = 10)


```

```{r}
## Statistical tests for Shannon diversity
#diversity$host_site_hostSection <- paste0(diversity$host, "_", diversity$site, "_", diversity$hostSection)
#shannon_pairwise_comparions <- pairwise.wilcox.test(diversity$shannon, diversity$host_site_hostSection, p.adjust.method = "bonferroni")
#write.csv(as.data.frame(shannon_pairwise_comparions$p.value), "./Outputs/shannonDiversityWilcoxonPairwiseTests-mangrove-PNG-16s.csv")
#shannon_pairwise_comparions

diversity$host_hostSection <- paste0(diversity$host, "_", diversity$hostSection)
shannon_pairwise_comparions <- pairwise.wilcox.test(diversity$shannon, diversity$host_hostSection, p.adjust.method = "bonferroni")
write.csv(as.data.frame(shannon_pairwise_comparions$p.value), "./Outputs/shannonDiversityWilcoxonPairwiseTests-mangrove-PNG-16s.csv")
shannon_pairwise_comparions

diversity$hostSection_site <- paste0(diversity$hostSection, "_", diversity$site)
shannon_pairwise_comparions <- pairwise.wilcox.test(diversity$shannon, diversity$hostSection_site, p.adjust.method = "bonferroni")
write.csv(as.data.frame(shannon_pairwise_comparions$p.value), "./Outputs/shannonDiversityWilcoxonPairwiseTests-hostSection_site-mangrove-PNG-16s.csv")
shannon_pairwise_comparions

diversity$hostSection_site_host <- paste0(diversity$hostSection, "_", diversity$site, "_", diversity$host)
shannon_pairwise_comparions <- pairwise.wilcox.test(diversity$shannon, diversity$hostSection_site_host, p.adjust.method = "bonferroni")
write.csv(as.data.frame(shannon_pairwise_comparions$p.value), "./Outputs/shannonDiversityWilcoxonPairwiseTests-hostSection_site_host-mangrove-PNG-16s.csv")

```

# Heatmap
## Mangrove

```{r}
library(microViz)
generateHeatmap <- function(psq, heatmap_rank = "Phylum", sample_arrange = "host"){
  psq <- prune_taxa(taxa_sums(psq) > 0, psq)
  htmap <- 
  psq %>% 
    tax_fix() %>% 
    # sort all samples by similarity
    ps_seriate(rank = heatmap_rank, tax_transform = "compositional", dist = "bray") %>% 
    # arrange the samples into Host, Host Section, Site groups
    ps_arrange(!!!syms(sample_arrange)) %>% 
    tax_transform("compositional", rank = heatmap_rank) %>%
    comp_heatmap(#sample_names_show = TRUE,
                 grid_col = NA,
                 #cluster_rows = FALSE,
                 sample_seriation = "Identity",
                 colors = heat_palette(breaks = c(0.0001, 0.001, 0.01, 0.1, 1), rev = T),
                 tax_anno = taxAnnotation(
                   Prev. = anno_tax_prev(bar_width = 0.3, size = grid::unit(1, "cm"))
                   ),
                 sample_anno = sampleAnnotation(
                   #"Host" = anno_sample("host"),
                   #"Host Section" = anno_sample("hostSection"),
                   #"Site" = anno_sample("site"),
                   #col = list(
                   #  "Host" = COLOR_HOST,
                   #  "Host Section" = COLOR_HOST_SECTION,
                   #  "Site" = COLOR_SITE
                   #  ), border = FALSE
                   "Host" = anno_sample_cat("host", col = COLOR_HOST, box_col = NA, border_col = "black", legend_title = "Host"),
                   "Host Section" = anno_sample_cat("hostSection", col = COLOR_HOST_SECTION, box_col = NA, border_col = "black", legend_title = "Host Section"),
                   "Site" = anno_sample_cat("site", col = COLOR_SITE, box_col = NA, border_col = "black", legend_title = "Site")
                   ),
                 heatmap_legend_param = list(at = c(0.0001, 0.001, 0.01, 0.1, 1), break_dist = 1)
                 )
  htmap %>% ComplexHeatmap::draw(
    annotation_legend_list = attr(htmap, "AnnoLegends")
    )
}
generateHeatmapTaxaSpecific <- function(psq, heatmap_rank = NA, sample_arrange = "host", taxaInclude = taxa_names(psq)){
  psq <- prune_taxa(taxa_sums(psq) > 0, psq)
  htmap <- 
    psq %>% 
    tax_fix() %>% 
    # sort all samples by similarity
    ps_seriate(rank = heatmap_rank, tax_transform = "compositional", dist = "bray") %>% 
    # arrange the samples into Host, Host Section, Site groups
    ps_arrange(!!!syms(sample_arrange)) %>% 
    tax_transform("compositional", rank = heatmap_rank) %>%
    comp_heatmap(#sample_names_show = TRUE,
                 taxa = taxaInclude,
                 taxon_renamer = renameTaxafromASV,
                 grid_col = NA,
                 #cluster_rows = FALSE,
                 sample_seriation = "Identity",
                 colors = heat_palette(breaks = c(0.0001, 0.001, 0.01, 0.1, 1), rev = T),
                 tax_anno = taxAnnotation(
                   Prev. = anno_tax_prev(bar_width = 0.3, size = grid::unit(1, "cm"))
                   ),
                 sample_anno = sampleAnnotation(
                   #"Host" = anno_sample("host"),
                   #"Host Section" = anno_sample("hostSection"),
                   #"Site" = anno_sample("site"),
                   #col = list(
                   #  "Host" = COLOR_HOST,
                   #  "Host Section" = COLOR_HOST_SECTION,
                   #  "Site" = COLOR_SITE
                   #  ), border = FALSE
                   "Host" = anno_sample_cat("host", col = COLOR_HOST, box_col = NA, border_col = "black", legend_title = "Host"),
                   "Host Section" = anno_sample_cat("hostSection", col = COLOR_HOST_SECTION, box_col = NA, border_col = "black", legend_title = "Host Section"),
                   "Site" = anno_sample_cat("site", col = COLOR_SITE, box_col = NA, border_col = "black", legend_title = "Site")
                   ),
                 heatmap_legend_param = list(at = c(0.0001, 0.001, 0.01, 0.1, 1), break_dist = 1)
                 )
  htmap %>% ComplexHeatmap::draw(
    annotation_legend_list = attr(htmap, "AnnoLegends")
    )
}

renameTaxafromASV <- function(ASVID, psq = ps.png.bac.mangrove){
  psq <- psq %>% tax_fix()
  renamedTaxa <- NULL
  for (i in ASVID){
    genus <- as.character(tax_table(psq)[c(i),c("Genus")])
    species <- as.character(tax_table(psq)[c(i),c("Species")])
    if (identical(genus, species)){
      renamedTaxa <- append(renamedTaxa, paste(str_split(genus, " ")[[1]][2], str_split(genus, " ")[[1]][1]))
    }else if(str_detect(species, "Genus")){
      renamedTaxa <- append(renamedTaxa, paste(str_split(species, " ")[[1]][2], str_split(species, " ")[[1]][1]))
    }else{
      renamedTaxa <- append(renamedTaxa, paste(genus, species))
    }
  }
  return(renamedTaxa)
}
```

```{r}
svglite("./Outputs/relativeAbundanceHeatmap-mangrove-phylum-PNG-16S.svg")
generateHeatmap(ps.png.bac.mangrove, heatmap_rank = "Phylum", sample_arrange = c("hostSection", "host", "site"))
dev.off()
svglite("./Outputs/relativeAbundanceHeatmap-mangrove-class-PNG-16S.svg")
generateHeatmap(ps.png.bac.mangrove, heatmap_rank = "Class", sample_arrange = c("hostSection", "host", "site"))
dev.off()
```

```{r}
# Checking contaminated samples
psra.png.bac.contams

htmap <- 
  psra.png.bac.contams %>% 
    tax_fix() %>% 
    # sort all samples by similarity
    ps_seriate(rank = "Class", tax_transform = "compositional", dist = "bray") %>% 
    # arrange the samples into Host, Host Section, Site groups
    ps_arrange(!!!syms(c("hostSection", "host"))) %>% 
    tax_transform("identity", rank = "Class") %>%
    comp_heatmap(#sample_names_show = TRUE,
                 grid_col = NA,
                 #cluster_rows = FALSE,
                 sample_seriation = "Identity",
                 tax_anno = taxAnnotation(
                   Prev. = anno_tax_prev(bar_width = 0.3, size = grid::unit(1, "cm"))
                   ),
                 sample_anno = sampleAnnotation(
                   #"Host" = anno_sample("host"),
                   #"Host Section" = anno_sample("hostSection"),
                   #"Site" = anno_sample("site"),
                   #col = list(
                   #  "Host" = COLOR_HOST,
                   #  "Host Section" = COLOR_HOST_SECTION,
                   #  "Site" = COLOR_SITE
                   #  ), border = FALSE
                   "Host" = anno_sample_cat("host", col = COLOR_HOST, box_col = NA, border_col = "black", legend_title = "Host"),
                   "Host Section" = anno_sample_cat("hostSection", col = COLOR_HOST_SECTION, box_col = NA, border_col = "black", legend_title = "Host Section"),
                   "Site" = anno_sample_cat("site", col = COLOR_SITE, box_col = NA, border_col = "black", legend_title = "Site")
                   )#,
                 #heatmap_legend_param = list(
                  # labels = rev(c("100%", " 10%", "  1%", " 0.1%", "0.01%"))
                  # )
                 )
htmap %>% ComplexHeatmap::draw(
  annotation_legend_list = attr(htmap, "AnnoLegends")
)
```


## Previous Method

```{r}
melted_ps.bac.mangrove <- psmelt(psra.png.bac.mangrove)
```

### By Phylum
```{r}
factor_melted_ps.bac.mangrove_by_siteHostSpecies <- unique(melted_ps.bac.mangrove[,c("site", "host", "hostSection")])
dataframe <- data.frame()  #creating dataframe to fill in information
melted_ps.bac.mangrove_aggregate <- aggregate(
  Abundance ~ OTU * site * host * hostSection,
  data = melted_ps.bac.mangrove,
  mean
)
tax.bac.mangrove_table <- as.data.frame(ps.png.bac.mangrove@tax_table)
tax.bac.mangrove_table$OTU <- row.names(tax.bac.mangrove_table)
melted_ps.bac.mangrove_aggregate <- left_join(melted_ps.bac.mangrove_aggregate,
                                     tax.bac.mangrove_table,
                                     by = "OTU")

for (i in 1:nrow(factor_melted_ps.bac.mangrove_by_siteHostSpecies)) {                                    #For each combination,
  sub_ps.bac.mangrove <- melted_ps.bac.mangrove_aggregate[melted_ps.bac.mangrove_aggregate$site == factor_melted_ps.bac.mangrove_by_siteHostSpecies[i,1] &
                                                          melted_ps.bac.mangrove_aggregate$host == factor_melted_ps.bac.mangrove_by_siteHostSpecies[i,2] &
                                                          melted_ps.bac.mangrove_aggregate$hostSection == factor_melted_ps.bac.mangrove_by_siteHostSpecies[i,3],] #subset dataframe.
  
  per_rank_abundance <- aggregate(Abundance ~ Phylum, data = sub_ps.bac.mangrove, sum)  #Aggregate abundance based on each rank
  
  per_rank_abundance$site <- sub_ps.bac.mangrove$site[1:nrow(per_rank_abundance)] #Adding sample name to each rank row
  per_rank_abundance$host <- sub_ps.bac.mangrove$host[1:nrow(per_rank_abundance)] 
  per_rank_abundance$hostSection <- sub_ps.bac.mangrove$hostSection[1:nrow(per_rank_abundance)]
  
  dataframe <- rbind(dataframe, per_rank_abundance)                       #store this in a dataframe for each row
}

#Sorting dataframe in order of hostTaxa, site, host, hostSection
ordered_df.mangrove <- dataframe[order(dataframe$site, dataframe$host, dataframe$hostSection),]
ordered_df.mangrove <- ordered_df.mangrove %>% 
  mutate(
    siteTag = case_when(
      site=="Kavieng" ~ "Kav", 
      site=="Kimbe Bay" ~ "Kim", 
      site=="Madang" ~ "Mad", 
      site=="Milne Bay" ~ "Mil", 
      site=="Port Moresby" ~ "Mot", 
      site=="Rabaul" ~ "Rab"
    ),
    hostTag = case_when(
      host=="Avicinea alba" ~ "Aa", 
      host=="Sonneratia alba" ~ "Sa"
    ),
    hostSectionTag = case_when(
      hostSection=="Fruit" ~ "Fru", 
      hostSection=="Leaf" ~ "Lea", 
      hostSection=="Pneumatophore" ~ "Pne", 
      hostSection=="Sediment" ~ "Sed"
    )
  )
ordered_df.mangrove$sortUID <- paste0(ordered_df.mangrove$hostSectionTag, ordered_df.mangrove$hostTag, ordered_df.mangrove$siteTag)
```

```{r}
table(ordered_df.mangrove$host) / length(unique(ordered_df.mangrove$Phylum))
heatplot <- ggplot(ordered_df.mangrove, aes(reorder(sortUID, sortUID), reorder(Phylum, Phylum))) +
  geom_tile(aes(fill = Abundance)) + 
  labs(y = "Phylum", x = "") + 
  scale_x_discrete(expand = c(0,0)) + 
  scale_y_discrete(expand = c(0,0)) +
  theme(axis.title = element_text(size = 20), axis.text.y = element_text(size = 13), 
        axis.text.x = element_blank(),#element_text(angle = 90), 
        axis.ticks.x = element_blank(),
        legend.text = element_text(size = 13), legend.title = element_text(size = 15),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  scale_fill_gradient2(
        #colors = c("white", "#132B43"),
        low = "white",
        mid = "#a6bddb",
        high = "#2b8cbe",#"#c62828","#132B43""#56B1F7""#1565c0""#ece7f2"
        midpoint = mean(ordered_df$Abundance),
        transform = "log",
        na.value = "white",
        breaks = 10^(-2*(5:0)),
        limits = c(sort(unique(ordered_df$Abundance))[2],1)
        )
  
y.min <- -0.5
y.max <- 0.5
y.mid <- y.min + (y.max - y.min) / 2
heatplot <- heatplot +
  
  # Host Taxa 1: Mangrove
  annotate("rect", xmin = 0.5, xmax = 9.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Fruit"]) +
  annotate("rect", xmin = 9.5, xmax = 18.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Leaf"]) +
  annotate("rect", xmin = 18.5, xmax = 27.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Pneumatophore"]) +
  annotate("rect", xmin = 27.5, xmax = 36.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST_SECTION["Sediment"]) + 
  annotate("text", x = 0.5 + (9.5 - 0.5) / 2, y = y.mid, label = names(COLOR_HOST_SECTION["Fruit"]), size = 3, srt = 0) +
  annotate("text", x = 9.5 + (18.5 - 9.5) / 2, y = y.mid, label = names(COLOR_HOST_SECTION["Leaf"]), size = 3, srt = 0) +
  annotate("text", x = 18.5 + (27.5 - 18.5) / 2, y = y.mid, label = names(COLOR_HOST_SECTION["Pneumatophore"]), size = 3, srt = 0) +
  annotate("text", x = 27.5 + (36.5 - 27.5) / 2, y = y.mid, label = names(COLOR_HOST_SECTION["Sediment"]), size = 3, srt = 0) +
  
  # For host section
  geom_vline(xintercept = c(9.5, 18.5, 27.5, 36.5, 48.5, 60.5, 72.5, 84.5), alpha = 0.2) +
  
  # Host Taxa
  annotate("rect", xmin = 0.5, xmax = 36.5, ymin = y.min - 1.5, ymax = y.min,
           alpha = 1, fill = COLOR_HOST_TAXA[1]) +
  annotate("text", x = 0.5 + (36.5 - 0.5) / 2, y = y.min - 1.5 / 2, label = names(COLOR_HOST_TAXA[1]), size = 5, srt = 0)

heatplot
ggsave(heatplot, filename = "./Outputs/relativeAbundanceHeatmap-mangrove-phylum-PNG-16s.svg", dpi = 300, width = 18, height = 15)

heatplot_linear <- heatplot + 
  scale_fill_gradient2(
        #colors = c("white", "#132B43"),
        low = "white",
        mid = "#a6bddb",
        high = "#2b8cbe",
        midpoint = 0.5,
        na.value = "white",
        limits = c(0,1)
        )
ggsave(heatplot_linear, filename = "./Outputs/relativeAbundanceHeatmapLinear-mangrove-phylum-PNG-16s.svg", dpi = 300, width = 18, height = 15)
```

### By Phylum - Individual samples
```{r}
factor_melted_ps.bac.mangrove_by_sample <- unique(melted_ps.bac.mangrove[,c("site", "host", "hostSection", "Sample")])
dataframe <- data.frame()  #creating dataframe to fill in information
tax.bac.mangrove_table <- as.data.frame(ps.png.bac.mangrove@tax_table)
tax.bac.mangrove_table$OTU <- row.names(tax.bac.mangrove_table)
melted_ps.bac.mangrove_by_sample <- melted_ps.bac.mangrove

for (i in 1:nrow(factor_melted_ps.bac.mangrove_by_sample)) {                                    #For each sample,
  sub_ps.bac.mangrove <- melted_ps.bac.mangrove_by_sample[melted_ps.bac.mangrove_by_sample$site == factor_melted_ps.bac.mangrove_by_sample[i,1] &
                                                          melted_ps.bac.mangrove_by_sample$host == factor_melted_ps.bac.mangrove_by_sample[i,2] &
                                                          melted_ps.bac.mangrove_by_sample$hostSection == factor_melted_ps.bac.mangrove_by_sample[i,3] &
                                                          melted_ps.bac.mangrove_by_sample$Sample == factor_melted_ps.bac.mangrove_by_sample[i,4],] #subset dataframe.
  
  per_rank_abundance <- aggregate(Abundance ~ Phylum, data = sub_ps.bac.mangrove, sum)  #Aggregate abundance based on each rank
  
  per_rank_abundance$site <- sub_ps.bac.mangrove$site[1:nrow(per_rank_abundance)] #Adding sample name to each rank row
  per_rank_abundance$host <- sub_ps.bac.mangrove$host[1:nrow(per_rank_abundance)] 
  per_rank_abundance$hostSection <- sub_ps.bac.mangrove$hostSection[1:nrow(per_rank_abundance)]
  per_rank_abundance$Sample <- sub_ps.bac.mangrove$Sample[1:nrow(per_rank_abundance)]
  
  dataframe <- rbind(dataframe, per_rank_abundance)                       #store this in a dataframe for each row
}

#Sorting dataframe in order of hostTaxa, site, host, hostSection
ordered_df.mangrove_by_sample <- dataframe[order(dataframe$site, dataframe$host, dataframe$hostSection),]
ordered_df.mangrove_by_sample <- ordered_df.mangrove_by_sample %>% 
  mutate(
    siteTag = case_when(
      site=="Kavieng" ~ "Kav", 
      site=="Kimbe Bay" ~ "Kim", 
      site=="Madang" ~ "Mad", 
      site=="Milne Bay" ~ "Mil", 
      site=="Port Moresby" ~ "Mot", 
      site=="Rabaul" ~ "Rab"
    ),
    hostTag = case_when(
      host=="Avicinea alba" ~ "Aa", 
      host=="Sonneratia alba" ~ "Sa"
    ),
    hostSectionTag = case_when(
      hostSection=="Fruit" ~ "Fru", 
      hostSection=="Leaf" ~ "Lea", 
      hostSection=="Pneumatophore" ~ "Pne", 
      hostSection=="Sediment" ~ "Sed"
    ),
    SampleTag = Sample
  )
ordered_df.mangrove_by_sample$sortUID <- paste0(ordered_df.mangrove_by_sample$hostSectionTag, ordered_df.mangrove_by_sample$hostTag, ordered_df.mangrove_by_sample$siteTag, ordered_df.mangrove_by_sample$Sample)
```

```{r}
table(ordered_df.mangrove_by_sample$host) / length(unique(ordered_df.mangrove$Phylum))
table(ordered_df.mangrove_by_sample$hostSection) / length(unique(ordered_df.mangrove$Phylum))
table(ordered_df.mangrove_by_sample$host, ordered_df.mangrove_by_sample$hostSection) / length(unique(ordered_df.mangrove$Phylum))
heatplot <- ggplot(ordered_df.mangrove_by_sample, aes(reorder(sortUID, sortUID), reorder(Phylum, Phylum))) +
  geom_tile(aes(fill = Abundance)) + 
  labs(y = "Phylum", x = "") + 
  scale_x_discrete(expand = c(0,0)) + 
  scale_y_discrete(expand = c(0,0)) +
  theme(axis.title = element_text(size = 20), axis.text.y = element_text(size = 13), 
        axis.text.x = element_blank(),#element_text(angle = 90), 
        axis.ticks.x = element_blank(),
        legend.text = element_text(size = 13), legend.title = element_text(size = 15),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  scale_fill_gradient2(
        #colors = c("white", "#132B43"),
        low = "white",
        mid = "#a6bddb",
        high = "#2b8cbe",#"#c62828","#132B43""#56B1F7""#1565c0""#ece7f2"
        midpoint = mean(ordered_df$Abundance),
        transform = "log",
        na.value = "white",
        breaks = 10^(-2*(5:0)),
        limits = c(sort(unique(ordered_df$Abundance))[2],1)
        )
  
y.min <- -0.5
y.max <- 0.5
y.mid <- y.min + (y.max - y.min) / 2
heatplot <- heatplot +
  
  # Host Species
  annotate("rect", xmin = 0.5, xmax = 40.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST["Avicinea alba"]) + 
  annotate("rect", xmin = 40.5, xmax = 100.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST["Sonneratia alba"]) + 
  annotate("rect", xmin = 100.5, xmax = 140.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST["Avicinea alba"]) + 
  annotate("rect", xmin = 140.5, xmax = 200.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST["Sonneratia alba"]) + 
  annotate("rect", xmin = 200.5, xmax = 240.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST["Avicinea alba"]) + 
  annotate("rect", xmin = 240.5, xmax = 300.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST["Sonneratia alba"]) + 
  annotate("rect", xmin = 300.5, xmax = 340.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST["Avicinea alba"]) + 
  annotate("rect", xmin = 340.5, xmax = 400.5, ymin = y.min, ymax = y.max,
           alpha = 0.5, fill = COLOR_HOST["Sonneratia alba"]) + 
  annotate("text", x = 0.5 + (40.5 - 0.5) / 2, y = y.mid, label = names(COLOR_HOST["Avicinea alba"]), size = 3, srt = 0) +
  annotate("text", x = 40.5 + (100.5 - 40.5) / 2, y = y.mid, label = names(COLOR_HOST["Sonneratia alba"]), size = 3, srt = 0) +
  annotate("text", x = 100.5 + (140.5 - 100.5) / 2, y = y.mid, label = names(COLOR_HOST["Avicinea alba"]), size = 3, srt = 0) +
  annotate("text", x = 140.5 + (200.5 - 140.5) / 2, y = y.mid, label = names(COLOR_HOST["Sonneratia alba"]), size = 3, srt = 0) +
  annotate("text", x = 200.5 + (240.5 - 200.5) / 2, y = y.mid, label = names(COLOR_HOST["Avicinea alba"]), size = 3, srt = 0) +
  annotate("text", x = 240.5 + (300.5 - 240.5) / 2, y = y.mid, label = names(COLOR_HOST["Sonneratia alba"]), size = 3, srt = 0) +
  annotate("text", x = 300.5 + (340.5 - 300.5) / 2, y = y.mid, label = names(COLOR_HOST["Avicinea alba"]), size = 3, srt = 0) +
  annotate("text", x = 340.5 + (400.5 - 340.5) / 2, y = y.mid, label = names(COLOR_HOST["Sonneratia alba"]), size = 3, srt = 0) +
  
  # For host species
  geom_vline(xintercept = c(40.5, 100.5, 140.5, 200.5, 240.5, 300.5, 340.5), alpha = 0.2) +
  
  # Host Section
  annotate("rect", xmin = 0.5, xmax = 100.5, ymin = y.min - 1.5, ymax = y.min,
           alpha = 1, fill = COLOR_HOST_SECTION["Fruit"]) +
  annotate("rect", xmin = 100.5, xmax = 200.5, ymin = y.min - 1.5, ymax = y.min,
           alpha = 1, fill = COLOR_HOST_SECTION["Leaf"]) +
  annotate("rect", xmin = 200.5, xmax = 300.5, ymin = y.min - 1.5, ymax = y.min,
           alpha = 1, fill = COLOR_HOST_SECTION["Pneumatophore"]) +
  annotate("rect", xmin = 300.5, xmax = 400.5, ymin = y.min - 1.5, ymax = y.min,
           alpha = 1, fill = COLOR_HOST_SECTION["Sediment"]) + 
  annotate("text", x = 0.5 + (100.5 - 0.5) / 2, y = y.min - 1.5 / 2, label = names(COLOR_HOST_SECTION["Fruit"]), size = 3, srt = 0) +
  annotate("text", x = 100.5 + (200.5 - 100.5) / 2, y = y.min - 1.5 / 2, label = names(COLOR_HOST_SECTION["Leaf"]), size = 3, srt = 0) +
  annotate("text", x = 200.5 + (300.5 - 200.5) / 2, y = y.min - 1.5 / 2, label = names(COLOR_HOST_SECTION["Pneumatophore"]), size = 3, srt = 0) +
  annotate("text", x = 300.5 + (400.5 - 300.5) / 2, y = y.min - 1.5 / 2, label = names(COLOR_HOST_SECTION["Sediment"]), size = 3, srt = 0) +
  
  # For host section
  geom_vline(xintercept = c(100.5, 200.5, 300.5), alpha = 0.5) +
  
  # Host Taxa
  annotate("rect", xmin = 0.5, xmax = 400.5, ymin = y.min - 2.5, ymax = y.min - 1,
           alpha = 1, fill = COLOR_HOST_TAXA[1]) +
  annotate("text", x = 0.5 + (400.5 - 0.5) / 2, y = y.min - 2.5 + (1.5) / 2, label = names(COLOR_HOST_TAXA[1]), size = 5, srt = 0)
  

heatplot
ggsave(heatplot, filename = "./Outputs/relativeAbundanceHeatmap-mangrove-sample-phylum-PNG-16s.svg", dpi = 300, width = 18, height = 15)

heatplot_linear <- heatplot + 
  scale_fill_gradient2(
        #colors = c("white", "#132B43"),
        low = "white",
        mid = "#a6bddb",
        high = "#2b8cbe",
        midpoint = 0.5,
        na.value = "white",
        limits = c(0,1)
        )
ggsave(heatplot_linear, filename = "./Outputs/relativeAbundanceHeatmapLinear-mangrove-sample-phylum-PNG-16s.svg", dpi = 300, width = 18, height = 15)
```

# Ordinations
```{r}
## Functions
veganCovEllipse <- function(cov, center = c(0, 0), scale = 1, npoints = 100){
  theta <- (0:npoints) * 2 * pi/npoints
  Circle <- cbind(cos(theta), sin(theta))
  t(center + scale * t(Circle %*% chol(cov)))
}

makeps.nmds <- function(ps.variable){
  ps.variable.nmds <- prune_samples(sample_sums(ps.variable) > 0, ps.variable)
  ps.variable.nmds <- prune_taxa(taxa_sums(ps.variable.nmds) > 0, ps.variable.nmds)
  ps.variable.nmds <- transform_sample_counts(ps.variable.nmds, function(otu){otu/sum(otu)})
  ps.variable.nmds
}

makeNMDSSimple <- function(ps.variable, label.variable, file.variable){
  ps.variable.nmds <- prune_samples(sample_sums(ps.variable) > 0, ps.variable)
  ps.variable.nmds <- prune_taxa(taxa_sums(ps.variable.nmds) > 0, ps.variable.nmds)
  ps.variable.nmds <- transform_sample_counts(ps.variable.nmds, function(otu){otu/sum(otu)})
  
  NMDS = ordinate(ps.variable.nmds, method = "NMDS", distance = "bray", try = 50, trymax = 1000)
  PCoA = ordinate(ps.variable.nmds, method = "PCoA", distance = "bray", try = 50, trymax = 1000)
  
  # NMDS with/without ellipses
  NMDS2 = data.frame(NMDS1 = NMDS$points[,1], 
                     NMDS2 = NMDS$points[,2],
                     site = ps.variable.nmds@sam_data$site,
                     host = ps.variable.nmds@sam_data$host,
                     hostTaxa = ps.variable.nmds@sam_data$hostTaxa,
                     hostSection = ps.variable.nmds@sam_data$hostSection)
  NMDS2.site.mean = aggregate(NMDS2[,1:2], list(site = ps.variable.nmds@sam_data$site), mean)
  NMDS2.host.mean = aggregate(NMDS2[,1:2], list(host = ps.variable.nmds@sam_data$host), mean)
  NMDS2.hostTaxa.mean = aggregate(NMDS2[,1:2], list(host = ps.variable.nmds@sam_data$hostTaxa), mean)
  NMDS2.hostSection.mean = aggregate(NMDS2[,1:2], list(host = ps.variable.nmds@sam_data$hostSection), mean)
  NMDS2$site <- as.factor(NMDS2$site)
  NMDS2$host <- as.factor(NMDS2$host)
  NMDS2$hostTaxa <- as.factor(NMDS2$hostTaxa)
  NMDS2$hostSection <- as.factor(NMDS2$hostSection)
  
  df_ell_site <- data.frame()
  for (g in levels(NMDS2$site)){
    df_ell_site <- rbind(df_ell_site, cbind(as.data.frame(with(NMDS2[NMDS2$site == g,],
                                                               veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                      wt = rep(1/length(NMDS1), length(NMDS1))
                                                               )$cov,
                                                               center = c(mean(NMDS1), mean(NMDS2))
                                                               ))), site = g))
  }
  
  df_ell_host <- data.frame()
  for (g in levels(NMDS2$host)){
    df_ell_host <- rbind(df_ell_host, cbind(as.data.frame(with(NMDS2[NMDS2$host == g,],
                                                                veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                       wt = rep(1/length(NMDS1), length(NMDS1))
                                                                )$cov,
                                                                center = c(mean(NMDS1), mean(NMDS2))
                                                                ))), host = g))
  }
  
  df_ell_hostTaxa <- data.frame()
  for (g in levels(NMDS2$hostTaxa)){
    df_ell_hostTaxa <- rbind(df_ell_hostTaxa, cbind(as.data.frame(with(NMDS2[NMDS2$hostTaxa == g,],
                                                               veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                      wt = rep(1/length(NMDS1), length(NMDS1))
                                                               )$cov,
                                                               center = c(mean(NMDS1), mean(NMDS2))
                                                               ))), hostTaxa = g))
  }
  
  df_ell_hostSection <- data.frame()
  for (g in levels(NMDS2$hostSection)){
    df_ell_hostSection <- rbind(df_ell_hostSection, cbind(as.data.frame(with(NMDS2[NMDS2$hostSection == g,],
                                                               veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                      wt = rep(1/length(NMDS1), length(NMDS1))
                                                               )$cov,
                                                               center = c(mean(NMDS1), mean(NMDS2))
                                                               ))), hostSection = g))
  }
  
  ### Site
  p_NMDS_site <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = site), size = 4) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_SITE) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(color = "Site")
  
  p_NMDS_site_ell <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = site), size = 4) +
    stat_ellipse(aes(color = site)) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_SITE) + 
    guides(shape = guide_legend(override.aes = list(size = 5)), color = guide_legend(override.aes = list(size = 5))) + 
    labs(color = "Site")
  
  p_PCoA_site <- plot_ordination(ps.variable.nmds, PCoA, color = "site") +  
    scale_color_manual(values = COLOR_SITE) + geom_point(size = 3) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = "Site", title = paste(label.variable," - ", "PCoA", sep = ""))
  
  ### Host
  
  p_NMDS_host <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = host), size = 4) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(color = "Host")
  
  p_NMDS_host_ell <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = host), size = 4) +
    stat_ellipse(aes(color = host)) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST) + 
    guides(shape = guide_legend(override.aes = list(size = 5)), color = guide_legend(override.aes = list(size = 5))) + 
    labs(color = "Host")
  
  p_PCoA_host <- plot_ordination(ps.variable.nmds, PCoA, color = "host") +  
    scale_color_manual(values = COLOR_HOST) + geom_point(size = 3) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = "Host", title = paste(label.variable," - ", "PCoA", sep = ""))
  
  ### HostTaxa
  
  p_NMDS_hostTaxa <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = hostTaxa), size = 4) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST_TAXA) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(color = "Host Taxa")
  
  p_NMDS_hostTaxa_ell <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = hostTaxa), size = 4) +
    stat_ellipse(aes(color = hostTaxa)) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST_TAXA) + 
    guides(shape = guide_legend(override.aes = list(size = 5)), color = guide_legend(override.aes = list(size = 5))) + 
    labs(color = "Host Taxa")
  
  p_PCoA_hostTaxa <- plot_ordination(ps.variable.nmds, PCoA, color = "hostTaxa") +  
    scale_color_manual(values = COLOR_HOST_TAXA) + geom_point(size = 3) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = "Host Taxa", title = paste(label.variable," - ", "PCoA", sep = ""))
  
  ### HostSection
  
  p_NMDS_hostSection <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = hostSection), size = 4) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST_SECTION) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(color = "Host Section")
  
  p_NMDS_hostSection_ell <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = hostSection), size = 4) +
    stat_ellipse(aes(color = hostSection)) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST_SECTION) + 
    guides(shape = guide_legend(override.aes = list(size = 5)), color = guide_legend(override.aes = list(size = 5))) + 
    labs(color = "Host Section")
  
  p_PCoA_hostSection <- plot_ordination(ps.variable.nmds, PCoA, color = "hostSection") +  
    scale_color_manual(values = COLOR_HOST_SECTION) + geom_point(size = 3) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(color = "Host Section", title = paste(label.variable," - ", "PCoA", sep = ""))
  
  ggsave(p_NMDS_site, filename = paste0("./Outputs/NMDS-site-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS_site_ell, filename = paste0("./Outputs/NMDS-site-ell-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_PCoA_site, filename = paste0("./Outputs/PCoA-site-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  
  ggsave(p_NMDS_host, filename = paste0("./Outputs/NMDS-host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS_host_ell, filename = paste0("./Outputs/NMDS-host-ell-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_PCoA_host, filename = paste0("./Outputs/PCoA-host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  
  ggsave(p_NMDS_hostTaxa, filename = paste0("./Outputs/NMDS-hostTaxa-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS_hostTaxa_ell, filename = paste0("./Outputs/NMDS-hostTaxa-ell-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_PCoA_hostTaxa, filename = paste0("./Outputs/PCoA-hostTaxa-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  
  ggsave(p_NMDS_hostSection, filename = paste0("./Outputs/NMDS-hostSection-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS_hostSection_ell, filename = paste0("./Outputs/NMDS-hostSection-ell-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_PCoA_hostSection, filename = paste0("./Outputs/PCoA-hostSection-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  
  p_NMDS_site
  p_NMDS_site_ell
  p_PCoA_site
  p_NMDS_host
  p_NMDS_host_ell
  p_PCoA_host
  p_NMDS_hostTaxa
  p_NMDS_hostTaxa_ell
  p_PCoA_hostTaxa
  p_NMDS_hostSection
  p_NMDS_hostSection_ell
  p_PCoA_hostSection
}

makeNMDSSingleFull <- function(ps.variable, label.variable, file.variable){
  ps.variable.nmds <- prune_samples(sample_sums(ps.variable) > 0, ps.variable)
  ps.variable.nmds <- prune_taxa(taxa_sums(ps.variable.nmds) > 0, ps.variable.nmds)
  ps.variable.nmds <- transform_sample_counts(ps.variable.nmds, function(otu){otu/sum(otu)})
  
  NMDS = ordinate(ps.variable.nmds, method = "NMDS", distance = "bray", try = 50, trymax = 1000)
  PCoA = ordinate(ps.variable.nmds, method = "PCoA", distance = "bray", try = 50, trymax = 1000)
  
  # NMDS with/without ellipses
  NMDS2 = data.frame(NMDS1 = NMDS$points[,1], 
                     NMDS2 = NMDS$points[,2],
                     site = ps.variable.nmds@sam_data$site,
                     host = ps.variable.nmds@sam_data$host,
                     hostTaxa = ps.variable.nmds@sam_data$hostTaxa,
                     hostSection = ps.variable.nmds@sam_data$hostSection)
  NMDS2.site.mean = aggregate(NMDS2[,1:2], list(site = ps.variable.nmds@sam_data$site), mean)
  NMDS2.host.mean = aggregate(NMDS2[,1:2], list(host = ps.variable.nmds@sam_data$host), mean)
  NMDS2.hostTaxa.mean = aggregate(NMDS2[,1:2], list(host = ps.variable.nmds@sam_data$hostTaxa), mean)
  NMDS2.hostSection.mean = aggregate(NMDS2[,1:2], list(host = ps.variable.nmds@sam_data$hostSection), mean)
  NMDS2$site <- as.factor(NMDS2$site)
  NMDS2$host <- as.factor(NMDS2$host)
  NMDS2$hostTaxa <- as.factor(NMDS2$hostTaxa)
  NMDS2$hostSection <- as.factor(NMDS2$hostSection)
  
  df_ell_site <- data.frame()
  for (g in levels(NMDS2$site)){
    df_ell_site <- rbind(df_ell_site, cbind(as.data.frame(with(NMDS2[NMDS2$site == g,],
                                                               veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                      wt = rep(1/length(NMDS1), length(NMDS1))
                                                               )$cov,
                                                               center = c(mean(NMDS1), mean(NMDS2))
                                                               ))), site = g))
  }
  
  df_ell_host <- data.frame()
  for (g in levels(NMDS2$host)){
    df_ell_host <- rbind(df_ell_host, cbind(as.data.frame(with(NMDS2[NMDS2$host == g,],
                                                                veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                       wt = rep(1/length(NMDS1), length(NMDS1))
                                                                )$cov,
                                                                center = c(mean(NMDS1), mean(NMDS2))
                                                                ))), host = g))
  }
  
  df_ell_hostTaxa <- data.frame()
  for (g in levels(NMDS2$hostTaxa)){
    df_ell_hostTaxa <- rbind(df_ell_hostTaxa, cbind(as.data.frame(with(NMDS2[NMDS2$hostTaxa == g,],
                                                               veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                      wt = rep(1/length(NMDS1), length(NMDS1))
                                                               )$cov,
                                                               center = c(mean(NMDS1), mean(NMDS2))
                                                               ))), hostTaxa = g))
  }
  
  df_ell_hostSection <- data.frame()
  for (g in levels(NMDS2$hostSection)){
    df_ell_hostSection <- rbind(df_ell_hostSection, cbind(as.data.frame(with(NMDS2[NMDS2$hostSection == g,],
                                                               veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                      wt = rep(1/length(NMDS1), length(NMDS1))
                                                               )$cov,
                                                               center = c(mean(NMDS1), mean(NMDS2))
                                                               ))), hostSection = g))
  }
  
  p_NMDS1 <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = hostSection, shape = host), size = 4) +
    stat_ellipse(aes(linetype = host), color = "#FFFFFF00", level = 1) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST_SECTION) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(shape = "Host", color = "Host Section", linetype = "Host")
  
  p_NMDS1_ell1 <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + 
    stat_ellipse(aes(fill = hostSection), color = "#FFFFFF00", level = 0.95, geom = "polygon", show.legend = FALSE) +
    geom_point(aes(color = hostSection, shape = host), size = 4) +
    stat_ellipse(aes(color = hostSection, linetype = host)) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST_SECTION) + 
    scale_fill_manual(values = scales::alpha(COLOR_HOST_SECTION, alpha = 0.10)) +
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(shape = "Host", color = "Host Section", linetype = "Ellipse Host")
  
  p_NMDS1_ell2 <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = hostSection, shape = host), size = 4) +
    stat_ellipse(aes(linetype = host), color = "#FFFFFF00", level = 1) +
    stat_ellipse(aes(linetype = host)) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST_SECTION) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(shape = "Host", color = "Host Section", linetype = "Host")
  
  p_PCoA1 <- plot_ordination(ps.variable.nmds, PCoA, color = "hostSection", shape = "host") +  
    scale_color_manual(values = COLOR_HOST_SECTION) + geom_point(size = 3) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(shape = "Host", color = "Host Section", title = paste(label.variable," - ", "PCoA", sep = ""))
  
  p_NMDS3 <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = host, shape = host), size = 4) +
    stat_ellipse(aes(linetype = host), color = "#FFFFFF00", level = 1) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_HOST) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(shape = "Host", color = "Host", linetype = "Host")
  
  p_NMDS4 <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = site, shape = host), size = 4) +
    stat_ellipse(aes(linetype = host), color = "#FFFFFF00", level = 1) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_SITE) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(shape = "Host", color = "Site", linetype = "Host")
  
  
  ggsave(p_NMDS1, filename = paste0("./Outputs/NMDS-hostSection_host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS1_ell1, filename = paste0("./Outputs/NMDS-hostSection_host-ell_hostSection-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS1_ell2, filename = paste0("./Outputs/NMDS-hostSection_host-ell_host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_PCoA1, filename = paste0("./Outputs/PCoA-hostSection_host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS3, filename = paste0("./Outputs/NMDS-host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS4, filename = paste0("./Outputs/NMDS-site-host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  
  p_NMDS1
}

makeNMDSSingleHostSection <- function(ps.variable, label.variable, file.variable){
  ps.variable.nmds <- prune_samples(sample_sums(ps.variable) > 0, ps.variable)
  ps.variable.nmds <- prune_taxa(taxa_sums(ps.variable.nmds) > 0, ps.variable.nmds)
  ps.variable.nmds <- transform_sample_counts(ps.variable.nmds, function(otu){otu/sum(otu)})
  
  NMDS = ordinate(ps.variable.nmds, method = "NMDS", distance = "bray", try = 50, trymax = 1000)
  PCoA = ordinate(ps.variable.nmds, method = "PCoA", distance = "bray", try = 50, trymax = 1000)
  
  # NMDS with/without ellipses
  NMDS2 = data.frame(NMDS1 = NMDS$points[,1], 
                     NMDS2 = NMDS$points[,2],
                     site = ps.variable.nmds@sam_data$site,
                     host = ps.variable.nmds@sam_data$host,
                     hostTaxa = ps.variable.nmds@sam_data$hostTaxa,
                     hostSection = ps.variable.nmds@sam_data$hostSection)
  NMDS2.site.mean = aggregate(NMDS2[,1:2], list(site = ps.variable.nmds@sam_data$site), mean)
  NMDS2.host.mean = aggregate(NMDS2[,1:2], list(host = ps.variable.nmds@sam_data$host), mean)
  NMDS2.hostTaxa.mean = aggregate(NMDS2[,1:2], list(host = ps.variable.nmds@sam_data$hostTaxa), mean)
  NMDS2.hostSection.mean = aggregate(NMDS2[,1:2], list(host = ps.variable.nmds@sam_data$hostSection), mean)
  NMDS2$site <- as.factor(NMDS2$site)
  NMDS2$host <- as.factor(NMDS2$host)
  NMDS2$hostTaxa <- as.factor(NMDS2$hostTaxa)
  NMDS2$hostSection <- as.factor(NMDS2$hostSection)
  
  df_ell_site <- data.frame()
  for (g in levels(NMDS2$site)){
    df_ell_site <- rbind(df_ell_site, cbind(as.data.frame(with(NMDS2[NMDS2$site == g,],
                                                               veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                      wt = rep(1/length(NMDS1), length(NMDS1))
                                                               )$cov,
                                                               center = c(mean(NMDS1), mean(NMDS2))
                                                               ))), site = g))
  }
  
  df_ell_host <- data.frame()
  for (g in levels(NMDS2$host)){
    df_ell_host <- rbind(df_ell_host, cbind(as.data.frame(with(NMDS2[NMDS2$host == g,],
                                                                veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                       wt = rep(1/length(NMDS1), length(NMDS1))
                                                                )$cov,
                                                                center = c(mean(NMDS1), mean(NMDS2))
                                                                ))), host = g))
  }
  
  df_ell_hostTaxa <- data.frame()
  for (g in levels(NMDS2$hostTaxa)){
    df_ell_hostTaxa <- rbind(df_ell_hostTaxa, cbind(as.data.frame(with(NMDS2[NMDS2$hostTaxa == g,],
                                                               veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                      wt = rep(1/length(NMDS1), length(NMDS1))
                                                               )$cov,
                                                               center = c(mean(NMDS1), mean(NMDS2))
                                                               ))), hostTaxa = g))
  }
  
  df_ell_hostSection <- data.frame()
  for (g in levels(NMDS2$hostSection)){
    df_ell_hostSection <- rbind(df_ell_hostSection, cbind(as.data.frame(with(NMDS2[NMDS2$hostSection == g,],
                                                               veganCovEllipse(cov.wt(cbind(NMDS1,NMDS2),
                                                                                      wt = rep(1/length(NMDS1), length(NMDS1))
                                                               )$cov,
                                                               center = c(mean(NMDS1), mean(NMDS2))
                                                               ))), hostSection = g))
  }
  
  p_NMDS2 <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = site, shape = host), size = 4) +
    stat_ellipse(aes(linetype = paste(host, site, sep = "_")), color = "#FFFFFF00") +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_SITE) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5)), linetype = "none") + 
    labs(shape = "Host", color = "Site", linetype = "Host")
  
  p_NMDS2_ell1 <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = site, shape = host), size = 4) +
    stat_ellipse(aes(linetype = host), color = "#FFFFFF00", level = 1) +
    stat_ellipse(aes(linetype = host)) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_SITE) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5))) + 
    labs(shape = "Host", color = "Site", linetype = "Ellipse Host")
  
  p_NMDS2_ell2 <- ggplot(data = NMDS2, aes(NMDS1, NMDS2)) + geom_point(aes(color = site, shape = host), size = 4) +
    stat_ellipse(aes(linetype = paste(host, site, sep = "_"))) +
    ggtitle(paste(label.variable," - ", "NMDS (Stress Value = ", toString(round(NMDS$stress, digits = 3)), ")", sep = "")) + 
    scale_color_manual(values = COLOR_SITE) + 
    theme(axis.title = element_text(size = 20), title = element_text(size = 20), 
          legend.text = element_text(size = 20)) + 
    guides(color = guide_legend(override.aes = list(size = 5)), linetype = "none") + 
    labs(shape = "Host", color = "Site")
  
  p_PCoA2 <- plot_ordination(ps.variable.nmds, PCoA, color = "site", shape = "host") +  
    scale_color_manual(values = COLOR_SITE) + geom_point(size = 3) +
    guides(color = guide_legend(override.aes = list(size = 4))) +
    labs(shape = "Host", color = "Site", title = paste(label.variable," - ", "PCoA", sep = ""))
  
  
  ggsave(p_NMDS2, filename = paste0("./Outputs/NMDS-site_host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS2_ell1, filename = paste0("./Outputs/NMDS-site_host-ell_host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_NMDS2_ell2, filename = paste0("./Outputs/NMDS-site_host-ell_host_site-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  ggsave(p_PCoA2, filename = paste0("./Outputs/PCoA-site_host-", file.variable, ".svg"), dpi = 300, width = 12, height = 10)
  
  p_NMDS2
}
```

```{r}
## Ordination Calculations and Plots
#makeNMDSSimple(ps.variable = ps.png.bac, label.variable = "PNG - 16S", file.variable = "PNG-16s")

#ps.png.bac.woSA_Mot_Pn_5 <- prune_samples(ps.png.bac@sam_data$sampleName != "SA_Mot_Pn_5", ps.png.bac)

#makeNMDSSimple(ps.variable = ps.png.bac.woSA_Mot_Pn_5, label.variable = "PNG without SA_Mot_Pn_5 - 16S", file.variable = "PNG-16s-woSA_Mot_Pn_5")

makeNMDSSimple(ps.variable = ps.png.bac.mangrove, label.variable = "Mangrove - PNG - 16S", file.variable = "mangrove-PNG-16s")

makeNMDSSimple(ps.variable = ps.png.bac.aa, label.variable = "Avicinea alba - PNG - 16S", file.variable = "aa-PNG-16s")
makeNMDSSimple(ps.variable = ps.png.bac.sa, label.variable = "Sonneratia alba - PNG - 16S", file.variable = "sa-PNG-16s")

makeNMDSSimple(ps.variable = ps.png.bac.mangrove.fruit, label.variable = "Mangrove Fruit - PNG - 16S", file.variable = "mangroveFruit-PNG-16s")
makeNMDSSimple(ps.variable = ps.png.bac.mangrove.leaf, label.variable = "Mangrove Leaf - PNG - 16S", file.variable = "mangroveLeaf-PNG-16s")
makeNMDSSimple(ps.variable = ps.png.bac.mangrove.pneumatophore, label.variable = "Mangrove Pneumatophore - PNG - 16S", file.variable = "mangrovePneumatophore-PNG-16s")
makeNMDSSimple(ps.variable = ps.png.bac.mangrove.sediment, label.variable = "Mangrove Sediment - PNG - 16S", file.variable = "mangroveSediment-PNG-16s")

makeNMDSSingleFull(ps.variable = ps.png.bac.mangrove, label.variable = "Mangrove - PNG - 16S", file.variable = "mangrove-PNG-16s")
makeNMDSSingleHostSection(ps.variable = ps.png.bac.mangrove.fruit, label.variable = "Mangrove Fruit - PNG - 16S", file.variable = "mangroveFruit-PNG-16s")
makeNMDSSingleHostSection(ps.variable = ps.png.bac.mangrove.leaf, label.variable = "Mangrove Leaf - PNG - 16S", file.variable = "mangroveLeaf-PNG-16s")
makeNMDSSingleHostSection(ps.variable = ps.png.bac.mangrove.pneumatophore, label.variable = "Mangrove Pneumatophore - PNG - 16S", file.variable = "mangrovePneumatophore-PNG-16s")
makeNMDSSingleHostSection(ps.variable = ps.png.bac.mangrove.sediment, label.variable = "Mangrove Sediment - PNG - 16S", file.variable = "mangroveSediment-PNG-16s")

# Excluding sediment samples:
makeNMDSSingleFull(ps.variable = ps.png.bac.mangrove.nosediment, label.variable = "Mangrove Excluding Sediment - PNG - 16S", file.variable = "mangroveNoSediment-PNG-16S")
```

# PERMANOVA
```{r}
ps.png.bac.mangrove.permanova <- prune_samples(sample_sums(ps.png.bac.mangrove) > 0, ps.png.bac.mangrove)
ps.png.bac.mangrove.permanova <- prune_taxa(taxa_sums(ps.png.bac.mangrove.permanova) > 0, ps.png.bac.mangrove.permanova)
ps.png.bac.mangrove.permanova <- transform_sample_counts(ps.png.bac.mangrove.permanova, function(otu){otu/sum(otu)})
otu = as.data.frame(otu_table(ps.png.bac.mangrove.permanova, taxa_are_rows = FALSE))
meta = as.data.frame(sample_data(ps.png.bac.mangrove.permanova))
df = data.frame(sampleName = meta$sampleName, site = meta$site, host = meta$host, hostTaxa = meta$hostTaxa, hostSection = meta$hostSection)

## Checking the Homogeneity Condition
#Note the assumption of similar multivariate spread among the groups, ie. analogous to variance homogeneity
#Permanova result may be potentially explained by significant different spreads within the groups themselves
dist <- vegdist(otu)
anova(betadisper(dist, meta$site))
anova(betadisper(dist, meta$hostSection))

## Test
permanova <- adonis2(otu ~ hostSection * site * host, data = df, by = "terms") #  * hostTaxa
sink("./Outputs/adonis_table-mangrove-PNG-16s.txt")
noquote("PERMANOVA for Mangrove Table:")
permanova
sink(NULL)

## Pairwise adonis
pairwise.adonis <- function(x,factors, sim.method = 'bray', p.adjust.m ='bonferroni'){
  co = combn(unique(factors),2)
  pairs = c()
  F.Model =c()
  R2 = c()
  p.value = c()
  for(elem in 1:ncol(co)){
    ad = adonis2(x[factors %in% c(co[1,elem],co[2,elem]),] ~ factors[factors %in% c(co[1,elem],co[2,elem])] , method =sim.method)
    pairs = c(pairs,paste(co[1,elem],'vs',co[2,elem]))
    F.Model =c(F.Model,ad$F[1])
    R2 = c(R2,ad$R2[1])
    p.value = c(p.value,ad$`Pr(>F)`[1])
  }
  p.adjusted = p.adjust(p.value,method=p.adjust.m)
  pairw.res = data.frame(pairs,F.Model,R2,p.value,p.adjusted)
  return(pairw.res)
}

### Site
padonis_site <- pairwise.adonis(otu, as.character(meta$site))
sink("./Outputs/padonis_site-aa-PNG-16s.txt")
noquote("Pairwise adonis between sites (Bonferroni corrected Pvalues):")
sink(NULL)
write.table(padonis_site, "./Outputs/padonis_site-mangrove-PNG-16s.txt", sep = "\t", quote = FALSE, append = TRUE)

### Host Section
padonis_hostSection <- pairwise.adonis(otu, as.character(meta$hostSection))
sink("./Outputs/padonis_hostSection-aa-PNG-16s.txt")
noquote("Pairwise adonis between host sections (Bonferroni corrected Pvalues):")
sink(NULL)
write.table(padonis_hostSection, "./Outputs/padonis_hostSection-mangrove-PNG-16s.txt", sep = "\t", quote = FALSE, append = TRUE)

###Look into notes from class on permanova and how to check assumptions###
```

```{r}
# Without sediment samples
ps.png.bac.mangrove.permanova <- prune_samples(sample_sums(ps.png.bac.mangrove.nosediment) > 0, ps.png.bac.mangrove.nosediment)
ps.png.bac.mangrove.permanova <- prune_taxa(taxa_sums(ps.png.bac.mangrove.permanova) > 0, ps.png.bac.mangrove.permanova)
ps.png.bac.mangrove.permanova <- transform_sample_counts(ps.png.bac.mangrove.permanova, function(otu){otu/sum(otu)})
otu = as.data.frame(otu_table(ps.png.bac.mangrove.permanova, taxa_are_rows = FALSE))
meta = as.data.frame(sample_data(ps.png.bac.mangrove.permanova))
df = data.frame(sampleName = meta$sampleName, site = meta$site, host = meta$host, hostTaxa = meta$hostTaxa, hostSection = meta$hostSection)

## Checking the Homogeneity Condition
#Note the assumption of similar multivariate spread among the groups, ie. analogous to variance homogeneity
#Permanova result may be potentially explained by significant different spreads within the groups themselves
dist <- vegdist(otu)
anova(betadisper(dist, meta$site))
anova(betadisper(dist, meta$hostSection))

## Test
permanova <- adonis2(otu ~ hostSection * site * host, data = df, by = "terms") #  * hostTaxa
sink("./Outputs/adonis_table-mangrove-noSediment-PNG-16s.txt")
noquote("PERMANOVA for Mangrove excluding Sediment Table:")
permanova
sink(NULL)
```

# Core Microbiome & Venn Diagrams

```{r}
library(microbiome)

# Atleast 10% relative abundance in atleast 20% of samples, across mangroves, host species, host sections.

core.mangrove <- core_members(ps.png.bac.mangrove, detection = 10/100, prevalence = 20/100)
core.mangrove.noSediment <- core_members(ps.png.bac.mangrove.nosediment, detection = 10/100, prevalence = 20/100)
core.aa <- core_members(ps.png.bac.aa, detection = 10/100, prevalence = 20/100)
core.aa.noSediment <- core_members(subset_samples(ps.png.bac.aa, hostSection != "Sediment"), detection = 10/100, prevalence = 20/100)
core.sa <- core_members(ps.png.bac.sa, detection = 10/100, prevalence = 20/100)
core.sa.noSediment <- core_members(subset_samples(ps.png.bac.sa, hostSection != "Sediment"), detection = 10/100, prevalence = 20/100)
core.mangroveFruit <- core_members(ps.png.bac.mangrove.fruit, detection = 10/100, prevalence = 20/100)
core.mangroveLeaf <- core_members(ps.png.bac.mangrove.leaf, detection = 10/100, prevalence = 20/100)
core.mangrovePneumatophore <- core_members(ps.png.bac.mangrove.pneumatophore, detection = 10/100, prevalence = 20/100)
core.mangroveSediment <- core_members(ps.png.bac.mangrove.sediment, detection = 10/100, prevalence = 20/100)

allCoreASVs <- sort(unique(c(core.mangrove, core.mangrove.noSediment, core.aa, core.sa, core.mangroveFruit, core.mangroveLeaf, core.mangrovePneumatophore, core.mangroveSediment)))
coreASVs <- data.frame(t(allCoreASVs))
colnames(coreASVs) <- coreASVs[1,]
coreASVs <- coreASVs[-1,]
coreASVs["ASVID",] <- allCoreASVs
coreASVs["Mangrove",] <- logical(ncol(coreASVs))
coreASVs["Mangrove excluding Sediment",] <- logical(ncol(coreASVs))
coreASVs["AA",] <- logical(ncol(coreASVs))
coreASVs["AA excluding Sediment",] <- logical(ncol(coreASVs))
coreASVs["SA",] <- logical(ncol(coreASVs))
coreASVs["SA excluding Sediment",] <- logical(ncol(coreASVs))
coreASVs["Mangrove Fruit",] <- logical(ncol(coreASVs))
coreASVs["Mangrove Leaf",] <- logical(ncol(coreASVs))
coreASVs["Mangrove Pneumatophore",] <- logical(ncol(coreASVs))
coreASVs["Mangrove Sediment",] <- logical(ncol(coreASVs))

coreASVs["Mangrove",core.mangrove] <- TRUE
coreASVs["Mangrove excluding Sediment",core.mangrove.noSediment] <- TRUE
coreASVs["AA",core.aa] <- TRUE
coreASVs["AA excluding Sediment",core.aa.noSediment] <- TRUE
coreASVs["SA",core.sa] <- TRUE
coreASVs["SA excluding Sediment",core.sa.noSediment] <- TRUE
coreASVs["Mangrove Fruit",core.mangroveFruit] <- TRUE
coreASVs["Mangrove Leaf",core.mangroveLeaf] <- TRUE
coreASVs["Mangrove Pneumatophore",core.mangrovePneumatophore] <- TRUE
coreASVs["Mangrove Sediment",core.mangroveSediment] <- TRUE
coreASVs <- data.frame(t(coreASVs))

coreTaxa <- data.frame(ASVID = allCoreASVs, data.frame(tax_table(ps.png.bac.mangrove)[allCoreASVs,]))

core_microbiome <- left_join(coreTaxa, coreASVs)

write.csv(core_microbiome, "./Outputs/core_microbiome-mangrove-PNG-16s.csv")

core_microbiome[core_microbiome$Mangrove == TRUE,]
core_microbiome[core_microbiome$Mangrove.excluding.Sediment == TRUE,]

length(core_microbiome[core_microbiome$AA == TRUE & core_microbiome$SA == TRUE,]$ASVID)
length(core_microbiome[core_microbiome$AA.excluding.Sediment == TRUE & core_microbiome$SA.excluding.Sediment == TRUE,]$ASVID)

svglite("./Outputs/coreMicrobiomeHeatmap-mangrove-PNG-16S.svg")
generateHeatmapTaxaSpecific(ps.png.bac.mangrove, sample_arrange = c("hostSection", "host", "site"), taxaInclude = core_microbiome[core_microbiome$AA == TRUE & core_microbiome$SA == TRUE,]$ASVID)
dev.off()

svglite("./Outputs/coreMicrobiomeHeatmap-mangrove_noSediment-PNG-16S.svg")
generateHeatmapTaxaSpecific(ps.png.bac.mangrove, sample_arrange = c("hostSection", "host", "site"), taxaInclude = core_microbiome[core_microbiome$AA.excluding.Sediment == TRUE & core_microbiome$SA.excluding.Sediment == TRUE,]$ASVID)
dev.off()
```

```{r}
# Venn Diagrams
library(MicEco)

venn_hostSection <- ps_venn(ps.png.bac.mangrove, "hostSection", fills = list(fill = COLOR_HOST_SECTION[c("Fruit", "Leaf", "Pneumatophore", "Sediment")], alpha = 0.5), quantities = list(type=c("percent","counts")))
venn_hostSection
svglite("./Outputs/vennDiagram-hostSection-mangrove-PNG-16S.svg")
venn_hostSection
dev.off()

venn_host <- ps_venn(ps.png.bac.mangrove, "host", fills = list(fill = COLOR_HOST[c("Avicinea alba", "Sonneratia alba")], alpha = 0.5), quantities = list(type=c("percent","counts")))
venn_host
svglite("./Outputs/vennDiagram-host-mangrove-PNG-16S.svg")
venn_host
dev.off()

venn_host_noSediment <- ps_venn(ps.png.bac.mangrove.nosediment, "host", fills = list(fill = COLOR_HOST[c("Avicinea alba", "Sonneratia alba")], alpha = 0.5), quantities = list(type=c("percent","counts")))
venn_host_noSediment
svglite("./Outputs/vennDiagram-host-mangrove_noSediment-PNG-16S.svg")
venn_host_noSediment
dev.off()

venn_hostSection_AA <- ps_venn(ps.png.bac.aa, "hostSection", fills = list(fill = COLOR_HOST_SECTION[c("Fruit", "Leaf", "Pneumatophore", "Sediment")], alpha = 0.5), quantities = list(type=c("percent","counts")))
venn_hostSection_AA

venn_hostSection_SA <- ps_venn(ps.png.bac.sa, "hostSection", fills = list(fill = COLOR_HOST_SECTION[c("Fruit", "Leaf", "Pneumatophore", "Sediment")], alpha = 0.5), quantities = list(type=c("percent","counts")))
venn_hostSection_SA
```


# Differential Abundance
```{r}
library(microViz)
```

## corncob
```{r}
library(corncob)
```

```{r}
# Creating summarized taxa names to plot
ps.png.bac.mangrove.DA <- prune_taxa(taxa_sums(ps.png.bac.mangrove) > 0, ps.png.bac.mangrove) %>% tax_fix()
tableNewNames <- data.frame(ASVID = rownames(tax_table(ps.png.bac.mangrove.DA)), tax_table(ps.png.bac.mangrove.DA)) %>% 
  mutate(
    Species = renameTaxafromASV(ASVID, psq = ps.png.bac.mangrove.DA) # Function from Heatmap section
  )
tax_table(ps.png.bac.mangrove.DA)[,7] <- tableNewNames$Species
```

```{r}
# Creating metadata field for host tissue versus sediment sectionSource

metaNew <- data.frame(sample_data(ps.png.bac.mangrove.DA))

metaNew <- metaNew %>% 
  mutate(
    sectionSource = case_when(
      hostSection == "Sediment" ~ "Sediment",
      .default = "Host Tissue"
    ),
    sectionSource = as.factor(sectionSource)
  )

sample_data(ps.png.bac.mangrove.DA) <- metaNew

psra.png.bac.mangrove.DA <- transform_sample_counts(ps.png.bac.mangrove.DA, function(otu){otu/sum(otu)})
```


```{r}
# Analysis for multiple taxa

## Differentially abundant taxa
set.seed(123)
sample_data(ps.png.bac.mangrove.DA) <- metaNew %>% 
  mutate(
    hostSection = factor(hostSection, levels = c("Sediment", "Pneumatophore", "Leaf", "Fruit"))
  )
hostSection_abun_analysis <- differentialTest(formula = ~ hostSection,
                                         phi.formula = ~ hostSection,
                                         formula_null = ~ 1,
                                         phi.formula_null = ~ hostSection,
                                         test = "Wald", boot = FALSE,
                                         data = ps.png.bac.mangrove.DA,
                                         fdr_cutoff = 0.05)
hostSection_abun_analysis
hostSection_abun_analysis$significant_taxa
#otu_to_taxonomy(OTU = hostSection_abun_analysis$significant_taxa, data = ps.png.bac.mangrove.DA)
diff_taxa_hostSection_corncob <- hostSection_abun_analysis$significant_taxa

site_abun_analysis <- differentialTest(formula = ~ site,
                                         phi.formula = ~ site,
                                         formula_null = ~ 1,
                                         phi.formula_null = ~ site,
                                         test = "Wald", boot = FALSE,
                                         data = ps.png.bac.mangrove.DA,
                                         fdr_cutoff = 0.05)
site_abun_analysis
site_abun_analysis$significant_taxa
diff_taxa_site_corncob <- site_abun_analysis$significant_taxa

host_abun_analysis <- differentialTest(formula = ~ host,
                                         phi.formula = ~ host,
                                         formula_null = ~ 1,
                                         phi.formula_null = ~ host,
                                         test = "Wald", boot = FALSE,
                                         data = ps.png.bac.mangrove.DA,
                                         fdr_cutoff = 0.05)
host_abun_analysis
host_abun_analysis$significant_taxa
diff_taxa_host_corncob <- host_abun_analysis$significant_taxa

sectionSource_abun_analysis <- differentialTest(formula = ~ sectionSource,
                                         phi.formula = ~ sectionSource,
                                         formula_null = ~ 1,
                                         phi.formula_null = ~ sectionSource,
                                         test = "Wald", boot = FALSE,
                                         data = ps.png.bac.mangrove.DA,
                                         fdr_cutoff = 0.05)
sectionSource_abun_analysis
sectionSource_abun_analysis$significant_taxa
#otu_to_taxonomy(OTU = sectionSource_abun_analysis$significant_taxa, data = ps.png.bac.mangrove.DA)
diff_taxa_sectionSource_corncob <- sectionSource_abun_analysis$significant_taxa

## Differentially variable taxa
#hostSection_var_analysis <- differentialTest(formula = ~ hostSection,
#                                             phi.formula = ~ hostSection,
#                                             formula_null = ~ hostSection,
#                                             phi.formula_null = ~ 1,
#                                             data = ps.png.bac.mangrove.DA,
#                                             test = "LRT", boot = FALSE,
#                                             fdr_cutoff = 0.05)
#hostSection_var_analysis
#hostSection_var_analysis$significant_taxa
#otu_to_taxonomy(OTU = hostSection_var_analysis$significant_taxa, data = ps.png.bac.mangrove.DA)
```

```{r}
plot(hostSection_abun_analysis, level = c("Genus"))
#plot(hostSection_var_analysis, level = c("Phylum"))
```

## ANCOM-BC2

```{r}
library(ANCOMBC)
library(tidyverse)
library(DT)
```

```{r}
# hostSection
set.seed(123)
output = ancombc2(data = psra.png.bac.mangrove, 
                             fix_formula = "host + hostSection + site", rand_formula = NULL,
                             p_adj_method = "holm", pseudo_sens = TRUE,
                             prv_cut = 0.01, lib_cut = 0, s0_perc = 0.05, # Settings changed for prv_cut to 1%
                             group = "hostSection", struc_zero = TRUE, neg_lb = TRUE,
                             alpha = 0.05, n_cl = 2, verbose = TRUE,
                             global = TRUE, pairwise = TRUE, 
                             dunnet = TRUE, trend = FALSE,
                             iter_control = list(tol = 1e-5, max_iter = 20, 
                                                 verbose = TRUE),
                             em_control = list(tol = 1e-5, max_iter = 100),
                             lme_control = lme4::lmerControl(), 
                             mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                             trend_control = NULL)

res_prim = output$res
res_pair = output$res_pair
res_global = output$res_global

diff_taxa_hostSection_ancombc <- res_global[res_global$passed_ss, "taxon"]
```

```{r}
# host
set.seed(123)
output = ancombc2(data = psra.png.bac.mangrove, 
                             fix_formula = "host + hostSection + site", rand_formula = NULL,
                             p_adj_method = "holm", pseudo_sens = TRUE,
                             prv_cut = 0.01, lib_cut = 0, s0_perc = 0.05, # Settings changed for prv_cut to 1%
                             group = "host", struc_zero = TRUE, neg_lb = TRUE,
                             alpha = 0.05, n_cl = 2, verbose = TRUE,
                             global = TRUE, pairwise = TRUE, 
                             dunnet = TRUE, trend = FALSE,
                             iter_control = list(tol = 1e-5, max_iter = 20, 
                                                 verbose = TRUE),
                             em_control = list(tol = 1e-5, max_iter = 100),
                             lme_control = lme4::lmerControl(), 
                             mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                             trend_control = NULL)

res_prim = output$res
res_pair = output$res_pair
res_global = output$res_global

diff_taxa_host_ancombc <- res_prim[res_prim$`passed_ss_hostSonneratia alba`, "taxon"]
```

```{r}
# site
set.seed(123)
output = ancombc2(data = psra.png.bac.mangrove, 
                             fix_formula = "host + hostSection + site", rand_formula = NULL,
                             p_adj_method = "holm", pseudo_sens = TRUE,
                             prv_cut = 0.01, lib_cut = 0, s0_perc = 0.05, # Settings changed for prv_cut to 1%
                             group = "site", struc_zero = TRUE, neg_lb = TRUE,
                             alpha = 0.05, n_cl = 2, verbose = TRUE,
                             global = TRUE, pairwise = TRUE, 
                             dunnet = TRUE, trend = FALSE,
                             iter_control = list(tol = 1e-5, max_iter = 20, 
                                                 verbose = TRUE),
                             em_control = list(tol = 1e-5, max_iter = 100),
                             lme_control = lme4::lmerControl(), 
                             mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                             trend_control = NULL)

res_prim = output$res
res_pair = output$res_pair
res_global = output$res_global

diff_taxa_site_ancombc <- res_global[res_global$passed_ss, "taxon"]
```

```{r}
# sectionSource
set.seed(123)
output = ancombc2(data = psra.png.bac.mangrove.DA, 
                             fix_formula = "host + site + sectionSource", rand_formula = NULL,
                             p_adj_method = "holm", pseudo_sens = TRUE,
                             prv_cut = 0.01, lib_cut = 0, s0_perc = 0.05, # Settings changed for prv_cut to 1%
                             group = "sectionSource", struc_zero = TRUE, neg_lb = TRUE,
                             alpha = 0.05, n_cl = 2, verbose = TRUE,
                             global = TRUE, pairwise = TRUE, 
                             dunnet = TRUE, trend = FALSE,
                             iter_control = list(tol = 1e-5, max_iter = 20, 
                                                 verbose = TRUE),
                             em_control = list(tol = 1e-5, max_iter = 100),
                             lme_control = lme4::lmerControl(), 
                             mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                             trend_control = NULL)

res = output$res

diff_taxa_sectionSource_ancombc <- res[res$passed_ss_sectionSourceSediment, "taxon"]
```

## IndicSpecies
```{r}
library(indicspecies)

otu.mangrove.ra <- data.frame(otu_table(psra.png.bac.mangrove))
meta.mangrove <- data.frame(sample_data(ps.png.bac.mangrove.DA))

indval_host <- multipatt(otu.mangrove.ra, meta.mangrove$host, func = "IndVal.g", 
                    control = how(nperm=999),
                    allow.negative = TRUE) 
summary(indval_host, indvalcomp=TRUE)
diff_taxa_host_indicspecies <- rownames(indval_host$sign[replace_na(indval_host$sign$p.value < 0.05, FALSE),])

indval_hostSection <- multipatt(otu.mangrove.ra, meta.mangrove$hostSection, func = "IndVal.g", 
                    control = how(nperm=999),
                    allow.negative = TRUE) 
summary(indval_hostSection, indvalcomp=TRUE)
diff_taxa_hostSection_indicspecies <- rownames(indval_hostSection$sign[replace_na(indval_hostSection$sign$p.value < 0.05, FALSE),])

indval_site <- multipatt(otu.mangrove.ra, meta.mangrove$site, func = "IndVal.g", 
                    control = how(nperm=999),
                    allow.negative = TRUE) 
summary(indval_site, indvalcomp=TRUE)
diff_taxa_site_indicspecies <- rownames(indval_site$sign[replace_na(indval_site$sign$p.value < 0.05, FALSE),])

indval_sectionSource <- multipatt(otu.mangrove.ra, meta.mangrove$sectionSource, func = "IndVal.g", 
                    control = how(nperm=999),
                    allow.negative = TRUE) 
summary(indval_sectionSource, indvalcomp=TRUE)
diff_taxa_sectionSource_indicspecies <- rownames(indval_sectionSource$sign[replace_na(indval_sectionSource$sign$p.value < 0.05, FALSE),])
```

## Aggregating Taxa that are identified using all three methods as differentially abundant

```{r}
# hostSection
diff_taxa_hostSection <- base::intersect(diff_taxa_hostSection_corncob, base::intersect(diff_taxa_hostSection_ancombc, diff_taxa_hostSection_indicspecies))
write.csv(diff_taxa_hostSection, file = "./Outputs/DAtaxalist_hostSection-mangrove-PNG-16S.csv")
length(diff_taxa_hostSection)

diff_taxa_hostSection <- read.csv("./Outputs/DAtaxalist_hostSection-mangrove-PNG-16S.csv")
diff_taxa_hostSection <- diff_taxa_hostSection$x

hostSection_DA_analysis <- hostSection_abun_analysis
hostSection_DA_analysis$significant_taxa <- hostSection_DA_analysis$significant_taxa[diff_taxa_hostSection_corncob %in% diff_taxa_hostSection]
hostSection_DA_analysis$significant_models <- hostSection_DA_analysis$significant_models[diff_taxa_hostSection_corncob %in% diff_taxa_hostSection]

plot(hostSection_DA_analysis, level = "Species")
svglite("./Outputs/DAtaxa_hostSection-mangrove-PNG-16S.svg")
plot(hostSection_DA_analysis, level = "Species") + labs(color = "Differentially\nabundant") + scale_color_manual(values = c("TRUE" = "#e6194b", "FALSE" = "#46f0f0")) + coord_flip()
dev.off()
```

```{r}
# Crucial Microbes
x = core_microbiome[core_microbiome$AA.excluding.Sediment == TRUE & core_microbiome$SA.excluding.Sediment == TRUE,]$ASVID

y = hostSection_DA_analysis$significant_taxa

z = base::intersect(x, y)

z

hostSection_crucial_analysis <- hostSection_abun_analysis
hostSection_crucial_analysis$significant_taxa <- hostSection_crucial_analysis$significant_taxa[diff_taxa_hostSection_corncob %in% z]
hostSection_crucial_analysis$significant_models <- hostSection_crucial_analysis$significant_models[diff_taxa_hostSection_corncob %in% z]

plot(hostSection_crucial_analysis, level = "Species")
svglite("./Outputs/DAtaxa_crucial-mangrove-PNG-16S.svg")
plot(hostSection_crucial_analysis, level = "Species") + labs(color = "Differentially\nabundant") + scale_color_manual(values = c("TRUE" = "#e6194b", "FALSE" = "#46f0f0"))
dev.off()

plot(hostSection_crucial_analysis, level = "Species") + labs(color = "Differentially\nabundant") + scale_color_manual(values = c("TRUE" = "#e6194b", "FALSE" = "#46f0f0"))
ggsave("./Outputs/crucial - DA plot -mangrove-PNG-16s.svg", width = 10, height = 4, dpi = 900)
```


```{r}
# host
diff_taxa_host <- base::intersect(diff_taxa_host_corncob, base::intersect(diff_taxa_host_ancombc, diff_taxa_host_indicspecies))
write.csv(diff_taxa_host, file = "./Outputs/DAtaxalist_host-mangrove-PNG-16S.csv")
length(diff_taxa_host)

host_DA_analysis <- host_abun_analysis
host_DA_analysis$significant_taxa <- host_DA_analysis$significant_taxa[diff_taxa_host_corncob %in% diff_taxa_host]
host_DA_analysis$significant_models <- host_DA_analysis$significant_models[diff_taxa_host_corncob %in% diff_taxa_host]

plot(host_DA_analysis, level = "Species")
svglite("./Outputs/DAtaxa_host-mangrove-PNG-16S.svg")
plot(host_DA_analysis, level = "Species") + labs(color = "Differentially\nabundant") + scale_color_manual(values = c("TRUE" = "#e6194b", "FALSE" = "#46f0f0"))
dev.off()
```

```{r}
# site
diff_taxa_site <- base::intersect(diff_taxa_site_corncob, base::intersect(diff_taxa_site_ancombc, diff_taxa_site_indicspecies))
write.csv(diff_taxa_site, file = "./Outputs/DAtaxalist_site-mangrove-PNG-16S.csv")
length(diff_taxa_site)

site_DA_analysis <- site_abun_analysis
site_DA_analysis$significant_taxa <- site_DA_analysis$significant_taxa[diff_taxa_site_corncob %in% diff_taxa_site]
site_DA_analysis$significant_models <- site_DA_analysis$significant_models[diff_taxa_site_corncob %in% diff_taxa_site]

plot(site_DA_analysis, level = "Species")
svglite("./Outputs/DAtaxa_site-mangrove-PNG-16S.svg")
plot(site_DA_analysis, level = "Species") + labs(color = "Differentially\nabundant") + scale_color_manual(values = c("TRUE" = "#e6194b", "FALSE" = "#46f0f0"))
dev.off()
```

```{r}
# sectionSource
diff_taxa_sectionSource <- base::intersect(diff_taxa_sectionSource_corncob, base::intersect(diff_taxa_sectionSource_ancombc, diff_taxa_sectionSource_indicspecies))
write.csv(diff_taxa_sectionSource, file = "./Outputs/DAtaxalist_sectionSource-mangrove-PNG-16S.csv")
length(diff_taxa_sectionSource)

sectionSource_DA_analysis <- sectionSource_abun_analysis
sectionSource_DA_analysis$significant_taxa <- sectionSource_DA_analysis$significant_taxa[diff_taxa_sectionSource_corncob %in% diff_taxa_sectionSource]

for(i in 1:length(sectionSource_DA_analysis$significant_models)){
  
}

sectionSource_DA_analysis$significant_models <- sectionSource_DA_analysis$significant_models[diff_taxa_sectionSource_corncob %in% diff_taxa_sectionSource]

sectionSource_DA_analysis$significant_taxa <- fct_reorder(sectionSource_DA_analysis$significant_taxa, )

plot(sectionSource_DA_analysis, level = "Species")
svglite("./Outputs/DAtaxa_sectionSource-mangrove-PNG-16S.svg")
plot(sectionSource_DA_analysis, level = "Species") + labs(color = "Differentially\nabundant") + scale_color_manual(values = c("TRUE" = "#e6194b", "FALSE" = "#46f0f0")) + coord_flip()
dev.off()

sectionSource_DA_plot <- plot(sectionSource_DA_analysis, level = "Species") + labs(color = "Differentially\nabundant") + scale_color_manual(values = c("TRUE" = "#e6194b", "FALSE" = "#46f0f0")) + coord_flip() + theme(axis.text.x = element_text(size = 5))
sectionSource_DA_plot
ggsave("./Outputs/sectionSource - DA plot -mangrove-PNG-16s.svg", width = 16, height = 9, dpi = 900)


# Sediment v Host Taxa & Core Host Taxa
x = core_microbiome[core_microbiome$AA.excluding.Sediment == TRUE & core_microbiome$SA.excluding.Sediment == TRUE,]$ASVID

y = sectionSource_DA_analysis$significant_taxa

z = base::intersect(x, y)

z

hostSediment_crucial_analysis <- sectionSource_abun_analysis
hostSediment_crucial_analysis$significant_taxa <- hostSediment_crucial_analysis$significant_taxa[diff_taxa_sectionSource_corncob %in% z]
hostSediment_crucial_analysis$significant_models <- hostSediment_crucial_analysis$significant_models[diff_taxa_sectionSource_corncob %in% z]

plot(hostSediment_crucial_analysis, level = "Species")
svglite("./Outputs/DAtaxa_crucialHostSediment-mangrove-PNG-16S.svg")
plot(hostSediment_crucial_analysis, level = "Species") + labs(color = "Differentially\nabundant") + scale_color_manual(values = c("TRUE" = "#e6194b", "FALSE" = "#46f0f0"))
dev.off()

plot(hostSediment_crucial_analysis, level = "Species") + labs(color = "Differentially\nabundant") + scale_color_manual(values = c("TRUE" = "#e6194b", "FALSE" = "#46f0f0"))
ggsave("./Outputs/crucial host sediment - DA plot -mangrove-PNG-16S.svg", width = 10, height = 4, dpi = 900)

```



# Spatial Analysis
```{r}
## Bio-ORACLE ##
#https://www.bio-oracle.org/code.php

#load packages
library(sdmpredictors)
library(leaflet)
library(raster)

# Explore datasets in the package
list_datasets()

# Explore layers in a dataset
list_layers(datasets = c("Bio-ORACLE", "MARSPEC"), marine = TRUE)
marine_layers <- list_layers(datasets = c("Bio-ORACLE", "MARSPEC"), marine = TRUE)
#write.csv(marine_layers,file="./Outputs/marine_layers_2025-08.csv",row.names = TRUE)

# Download specific layers to the current directory
options(sdmpredictors_datadir="./sdmpredictorsData")
env.layers <- load_layers(c("BO_calcite","BO_cloudmean","BO_damean","BO_parmean","BO_ph","BO21_salinitymean_ss","BO21_chlomean_ss","BO21_curvelmean_ss","BO21_dissoxmean_ss","BO21_nitratemean_ss","BO21_phosphatemean_ss","BO21_ppmean_ss","BO21_silicatemean_ss","BO21_tempmean_ss","BO21_temprange_ss"))

#Generate a data.frame with the sites of interest
gpsCoords <- read.csv("../../Resources/gpsCoords.csv")[,1:3]
gpsCoords$lon <- parzer::parse_lon(gpsCoords$lon)
gpsCoords$lat <- parzer::parse_lat(gpsCoords$lat)
rownames(gpsCoords) <- gpsCoords$site

my.sites <- gpsCoords[order(gpsCoords$site),]
my.sites

# Maps
library(marmap)
png.bathy <- getNOAA.bathy(lon1 = 143.0300, lon2 = 156.9167, lat1 =-11.9416, lat2 = -1.8114, 
                           resolution = 10, keep = TRUE)
plot(png.bathy)

# Extract environmental values from layers
#shape.raster <- raster("./maps/GMRTv4_3_1_20250825topo_PNG.grd")
#plot(shape.raster>0)
env.layers.df <- as.data.frame(env.layers, xy = TRUE) %>% na.omit()
ggplot() + geom_raster(data = env.layers.df, aes(x, y, fill = BO21_tempmean_ss)) +
  lims(
    x = c(143.0300, 156.9167),
    y = c(-11.9416, -1.8114)
  ) +
  geom_point(data = my.sites, aes(lon, lat, color = site)) +
  geom_label(data = my.sites, aes(lon, lat, label = site), nudge_y = 0.5) + 
  scale_color_manual(values = COLOR_SITE)

plot(png.bathy)
points(my.sites$lon, my.sites$lat, col = COLOR_SITE, pch = 17)

my.sites.environment <- data.frame(Name = my.sites$site, extract(env.layers, my.sites[,c("lon", "lat")]))
#write.csv(my.sites.environment,file="./sdmpredictorsData/PNG_env_extract_all.csv",row.names = TRUE)
row.names(my.sites.environment) <- my.sites.environment$Name
env_full <- my.sites.environment[meta$site,][,-1]
row.names(env_full) <- meta$sampleName
all_env_dist <- dist(as.matrix(env_full), method = "euclidean")
colnames(env_full)

```

## Mantel Test
```{r}
library(geosphere)

psra.png.bac.mantel <- psra.png.bac.mangrove

## Extracting Longitude and Latitude data
meta <- data.frame(sample_data(psra.png.bac.mantel))

my.sites
geo_full <- my.sites[meta$site,]
row.names(geo_full) <- meta$sampleName

## Preparing asv tables
otu_full <- data.frame(otu_table(psra.png.bac.mantel))

## Making distance matrices
# Adundance data frames - bray curtis dissimilarity
dist.otu_full <- vegdist(otu_full, method = "bray")

# Geographic data frame - haversie distance
d.geo_full <- distm(geo_full[,c("lon", "lat")], fun = distHaversine)

dist.geo_full <- as.dist(d.geo_full)

## Running Mantel Test
# Abundance vs Geographic
abund_geo_full <- vegan::mantel(dist.otu_full, dist.geo_full, method = "spearman", permutations = 9999, na.rm = TRUE)
abund_geo_full
## Saving Output
sink("./Outputs/Mantel_Test-mangrove-all-PNG-16S.txt")
noquote("Mantel Test on All Samples")
abund_geo_full
sink(NULL)

# For each host overall
i = base::unique(meta$host)[1] # set to desired host
i
geo_mantel <- geo_full[meta$host == i,]
otu_mantel <- otu_full[rownames(geo_mantel),]
dist.otu_mantel <- vegdist(otu_mantel, method = "bray")
d.geo_mantel <- distm(geo_mantel[,c("lon", "lat")], fun = distHaversine)
dist.geo_mantel <- as.dist(d.geo_mantel)
#Mantel
abund_geo_mantel <- vegan::mantel(dist.otu_mantel, dist.geo_mantel, method = "spearman", permutations = 9999, na.rm = TRUE)
abund_geo_mantel$statistic ^ 2
abund_geo_mantel$signif
#MRM
abund_geo_mrm <- ecodist::MRM(dist.otu_mantel ~ dist.geo_mantel,  nperm = 9999)
abund_geo_mrm$r.squared[1] ^ 2
abund_geo_mrm$r.squared[2]



# For each host in each section
sink("./Outputs/Mantel_Test-mangrove-PNG-16S.txt")
sink(NULL)
mantel_R2 <- data.frame(row.names = base::unique(meta$hostSection), hostSection = base::unique(meta$hostSection))
mantel_p <- data.frame(row.names = base::unique(meta$hostSection), hostSection = base::unique(meta$hostSection))
mrm_R2 <- data.frame(row.names = base::unique(meta$hostSection), hostSection = base::unique(meta$hostSection))
mrm_p <- data.frame(row.names = base::unique(meta$hostSection), hostSection = base::unique(meta$hostSection))
for (i in base::unique(meta$host)){
  mantel_R2[,eval(i)] <- numeric(length(base::unique(meta$hostSection)))
  mantel_p[,eval(i)] <- numeric(length(base::unique(meta$hostSection)))
  for (j in base::unique(meta$hostSection)){
    geo_mantel <- geo_full[meta$host == i & meta$hostSection == j,]
    otu_mantel <- otu_full[rownames(geo_mantel),]
    
    dist.otu_mantel <- vegdist(otu_mantel, method = "bray")
    d.geo_mantel <- distm(geo_mantel[,c("lon", "lat")], fun = distHaversine)
    dist.geo_mantel <- as.dist(d.geo_mantel)
    
    abund_geo_mantel <- vegan::mantel(dist.otu_mantel, dist.geo_mantel, method = "spearman", permutations = 9999, na.rm = TRUE)
    print(paste0("Mantel test on ", i, " samples in ", j, ":"))
    print(abund_geo_mantel)
    
    mantel_R2[j, i] = abund_geo_mantel$statistic ^ 2
    mantel_p[j, i] = abund_geo_mantel$signif
    
    abund_geo_mrm <- ecodist::MRM(dist.otu_mantel ~ dist.geo_mantel,  nperm = 9999)
    print(paste0("MRM on ", i, " samples in ", j, ":"))
    print(abund_geo_mrm)
    
    mrm_R2[j, i] = abund_geo_mrm$r.squared[1] ^ 2
    mrm_p[j, i] = abund_geo_mrm$r.squared[2]
  }
}# Manually copy paste output into file.
write.csv(mantel_R2, file = "./Outputs/Mantel_Test_R2-mangrove-PNG-16S.csv")
write.csv(mantel_p, file = "./Outputs/Mantel_Test_p-mangrove-PNG-16S.csv")
write.csv(mrm_R2, file = "./Outputs/MRM_R2-mangrove-PNG-16S.csv")
write.csv(mrm_p, file = "./Outputs/MRM_p-mangrove-PNG-16S.csv")
mantel_R2
mantel_p
mrm_R2
mrm_p

### Multiple Regression on distance matrices ###
library(ecodist)
dist_MRM <- ecodist::MRM(dist.otu_full ~ dist.geo_full + all_env_dist,  nperm = 9999)
dist_MRM

my.sites.environment
site_env_pca <- pca(my.sites.environment[2:ncol(my.sites.environment)], scale = TRUE)
plot(site_env_pca, arrows = TRUE)

identical(rownames(otu.mangrove.ra), rownames(env_full))
otu_env_rda <- rda(otu.mangrove.ra, env_full)
plot(otu_env_rda)

library(ggvegan)
autoplot(otu_env_rda, arrows = TRUE)

sink("./Outputs/Analysis/MRM_Table_16S.txt")
print("Bray-Curtis distance regressed against geographic distance (Multiple regression on matrices):")
ecodist::MRM(dist.otu_full ~ dist.geo_full,  nperm = 9999)

print("Bray-Curtis distance regressed against environmental distance (Multiple regression on matrices):")
ecodist::MRM(dist.otu_full ~ all_env_dist,  nperm = 9999)

print("Bray-Curtis distance regressed against geographic + environmental distance (Multiple regression on matrices):")
ecodist::MRM(dist.otu_full ~ dist.geo_full + all_env_dist,  nperm = 9999)
sink(NULL)
```


## Run Multicollinearity checks
```{r}
#Load packages
library(tidyverse)
library(psych)
library(adespatial)
library(vegan)
library(GGally)

# Import environmental data
env.raw = read.csv("./sdmpredictorsData/PNG_env_extract_all.csv", row.names = 1)

# Plot and run correlation test on environmental variables
pairs.panels(env.raw, scale = TRUE)

ggpairs(env.raw, columns = 1:15)
ggcorr(env.raw)
ggcorr(env.raw, geom = "circle",
       max_size = 10,
       size = 4,
       hjust = 1,
       layout.exp = 3,
       label = TRUE,
       label_size = 3,
       palette = "PuOr")
#ggsave("./Outputs/Correlation Test 1-aa-PNG-16s.svg", width = 15, height =12, dpi = 900)

ggcorr(env.raw, geom = "blank", 
       size = 8, 
       hjust = 0.75,
       layout.exp = 1,
       label = TRUE, 
       label_size = 8) +
   geom_point(size = 15, aes(color = coefficient > 0, alpha = abs(coefficient) > 0.5)) +
   scale_alpha_manual(values = c("TRUE" = 0.25, "FALSE" = 0)) +
   guides(color = "none", alpha = "none")
#ggsave("./Outputs/Correlation Test 2-aa-PNG-16s.svg", width = 15, height = 12, dpi = 900)

#Remove parameters with pairwise collinearity values greater that 0.7
# Correlations:
# parmean ~ damean
# salinitymean ~ parmean
# chlomean ~ ph
# dissoxmean ~ damean
# nitratemean ~ chlomean
# ph ~ nitratemean
# phosphatemean ~ damean
# phosphatemean ~ chlomean
# phosphatemean ~ dissoxmean
# ppmean ~ phosphatemean
# ppmean ~ nitratemean
# ppmean ~ chlomean
# ppmean ~ ph
# silicatemean ~ damean
# silicatemean ~ parmean
# silicatemean ~ salinitymean
# silicatemean ~ dissoxmean
# silicatemean ~ phosphatemean
# tempmean ~ dissoxmean
# tempmean ~ ph
# temprange ~ damean
# temprange ~ ph
# temprange ~ chlomean
# temprange ~ dissoxmean
# temprange ~ nitratemean
# temprange ~ phosphatmean
# temprange ~ ppmean
# temprange ~ tempmean

library(igraph)
highCorrelations <- data.frame(a = c("parmean", "salinitymean", "chlomean", "dissoxmean", "nitratemean", "ph", "phosphatemean", "phosphatemean", "phosphatemean", "ppmean", "ppmean", "ppmean", "ppmean", "silicatemean", "silicatemean", "silicatemean", "silicatemean", "silicatemean", "tempmean", "tempmean", "temprange", "temprange", "temprange", "temprange", "temprange", "temprange", "temprange", "temprange"), b = c("damean", "parmean", "ph", "damean", "chlomean", "nitratemean", "damean", "chlomean", "dissoxmean", "phosphatemean", "nitratemean", "chlomean", "ph", "damean", "parmean", "salinitymean", "dissoxmean", "phosphatemean", "dissoxmean", "ph", "damean", "ph", "chlomean", "dissoxmean", "nitratemean", "phosphatemean", "ppmean", "tempmean"))
graph <- graph_from_data_frame(highCorrelations, directed = FALSE)
plot(graph)
variables_to_remove <- c("temprange", "silicatemean", "ppmean", "chlomean", "dissoxmean", # Main
                         "ph", "damean", "parmean"  # Additional
                         ) 
graph_less <- graph_from_data_frame(highCorrelations[!(highCorrelations$a %in% variables_to_remove | highCorrelations$b %in% variables_to_remove),], directed = FALSE)
plot(graph_less)

colnames(env.raw)
variables_to_remove_colname <- c("BO21_ppmean_ss", "BO21_silicatemean_ss", "BO21_temprange_ss", "BO21_chlomean_ss", "BO21_dissoxmean_ss",
                                 "BO_ph", "BO_damean", "BO_parmean",
                                 "BO21_phosphatemean_ss", "BO_calcite", "BO21_curvelmean_ss"
                                 )
env.filtered <- env.raw[,setdiff(colnames(env.raw), variables_to_remove_colname)]
ggcorr(env.filtered, geom = "blank", size = 8, label = TRUE, label_size = 8, hjust = 0.75,layout.exp = 1) +
   geom_point(size = 15, aes(color = coefficient > 0, alpha = abs(coefficient) > 0.5)) +
   scale_alpha_manual(values = c("TRUE" = 0.25, "FALSE" = 0)) +
   guides(color = "none", alpha = "none")
ggsave("./Outputs/correlation test-aa-PNG-16sfiltered.svg", width = 15, height = 12, dpi = 900)

colnames(env.filtered)
```
      
## Ordinations  
```{r}
# NMDS for Env Parameters
aa_env <- dplyr::left_join(meta.aa, env.filtered, by = join_by(site == Name))
aa_env <- aa_env[,11:ncol(aa_env)]
aa_env_distmat <- vegdist(aa_env, method = "bray")
#aa_env_NMS <- metaMDS(aa_env_distmat, distance = "bray", k = 2, maxit = 999, trymax = 5000, wascores = TRUE)
#plot(aa_env_NMS, "sites")
#orditorp(aa_env_NMS,"sites")
#goodness(aa_env_NMS)

## load data - relative abundance (seq)
rel_seq <- data.frame(psra.png.bac.aa@otu_table)
meta_full <- dplyr::left_join(meta.aa, env.filtered, by = join_by(site == Name))

#Calculate distance matrix
rel_seq_distmat <- vegdist(rel_seq, method = "bray")
#Create easy to view matrix and write as .csv
rel_seq_distmat <- as.matrix(rel_seq_distmat, labels = T)
#write.csv(rel_seq_distmat, "aa_rel_seq_distmat.csv")

#running NMDS in vegan (metaMDS)
#k=2
rel_seq_NMS <- metaMDS(rel_seq_distmat, distance = "bray", k = 2, maxit = 999, trymax = 100, wascores = TRUE)
rel_seq_NMS

#Goodness of fit/Shepards test
goodness(rel_seq_NMS) #produces a result of test statistics for goodness of fit for each point
stressplot(rel_seq_NMS)

#plot NMDS
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
envfit <- envfit(rel_seq_NMS ~ BO_cloudmean + BO21_salinitymean_ss + BO21_nitratemean_ss + BO21_tempmean_ss,
                 data=meta_full,
                 perm=999)
#envfit <- envfit(rel_seq_NMS ~ BO_calcite + BO_cloudmean + BO_damean + BO_parmean + BO_ph + BO21_salinitymean_ss + BO21_chlomean_ss + BO21_curvelmean_ss + BO21_dissoxmean_ss + BO21_nitratemean_ss + BO21_phosphatemean_ss + BO21_ppmean_ss + BO21_silicatemean_ss + BO21_tempmean_ss + BO21_temprange_ss,data=meta_full,perm=999)
scores(envfit, "vectors")
plot(envfit)

svg("./Outputs/NMDS_w_env-aa-PNG-16s.svg")
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
plot(envfit)
dev.off()
```

## Running ordinations for each hostSection
### Fruit
```{r}
# NMDS for Env Parameters
## load data - relative abundance (seq)
rel_seq <- data.frame(psra.png.bac.aa.fruit@otu_table)
meta_full <- dplyr::left_join(data.frame(psra.png.bac.aa.fruit@sam_data), env.filtered, by = join_by(site == Name))

#Calculate distance matrix
rel_seq_distmat <- vegdist(rel_seq, method = "bray")
rel_seq_distmat <- as.matrix(rel_seq_distmat, labels = T)

#running NMDS in vegan (metaMDS)
#k=2
rel_seq_NMS <- metaMDS(rel_seq_distmat, distance = "bray", k = 2, maxit = 999, trymax = 100, wascores = TRUE)
rel_seq_NMS

#Goodness of fit/Shepards test
goodness(rel_seq_NMS) #produces a result of test statistics for goodness of fit for each point
stressplot(rel_seq_NMS)

#plot NMDS
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
envfit <- envfit(rel_seq_NMS ~ BO_cloudmean + BO21_salinitymean_ss + BO21_nitratemean_ss + BO21_tempmean_ss,
                 data=meta_full,
                 perm=999)
scores(envfit, "vectors")
plot(envfit)

svg("./Outputs/NMDS_w_env-aa-fruit-PNG-16s.svg")
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
plot(envfit)
dev.off()
```

### Leaf
```{r}
# NMDS for Env Parameters
## load data - relative abundance (seq)
rel_seq <- data.frame(psra.png.bac.aa.leaf@otu_table)
meta_full <- dplyr::left_join(data.frame(psra.png.bac.aa.leaf@sam_data), env.filtered, by = join_by(site == Name))

#Calculate distance matrix
rel_seq_distmat <- vegdist(rel_seq, method = "bray")
rel_seq_distmat <- as.matrix(rel_seq_distmat, labels = T)

#running NMDS in vegan (metaMDS)
#k=2
rel_seq_NMS <- metaMDS(rel_seq_distmat, distance = "bray", k = 2, maxit = 999, trymax = 100, wascores = TRUE)
rel_seq_NMS

#Goodness of fit/Shepards test
goodness(rel_seq_NMS) #produces a result of test statistics for goodness of fit for each point
stressplot(rel_seq_NMS)

#plot NMDS
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
envfit <- envfit(rel_seq_NMS ~ BO_cloudmean + BO21_salinitymean_ss + BO21_nitratemean_ss + BO21_tempmean_ss,
                 data=meta_full,
                 perm=999)
scores(envfit, "vectors")
plot(envfit)

svg("./Outputs/NMDS_w_env-aa-leaf-PNG-16s.svg")
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
plot(envfit)
dev.off()
```

### Pneumatophore
```{r}
# NMDS for Env Parameters
## load data - relative abundance (seq)
rel_seq <- data.frame(psra.png.bac.aa.pneumatophore@otu_table)
meta_full <- dplyr::left_join(data.frame(psra.png.bac.aa.pneumatophore@sam_data), env.filtered, by = join_by(site == Name))

#Calculate distance matrix
rel_seq_distmat <- vegdist(rel_seq, method = "bray")
rel_seq_distmat <- as.matrix(rel_seq_distmat, labels = T)

#running NMDS in vegan (metaMDS)
#k=2
rel_seq_NMS <- metaMDS(rel_seq_distmat, distance = "bray", k = 2, maxit = 999, trymax = 100, wascores = TRUE)
rel_seq_NMS

#Goodness of fit/Shepards test
goodness(rel_seq_NMS) #produces a result of test statistics for goodness of fit for each point
stressplot(rel_seq_NMS)

#plot NMDS
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
envfit <- envfit(rel_seq_NMS ~ BO_cloudmean + BO21_salinitymean_ss + BO21_nitratemean_ss + BO21_tempmean_ss,
                 data=meta_full,
                 perm=999)
scores(envfit, "vectors")
plot(envfit)

svg("./Outputs/NMDS_w_env-aa-pneumatophore-PNG-16s.svg")
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
plot(envfit)
dev.off()
```

### Sediment
```{r}
# NMDS for Env Parameters
## load data - relative abundance (seq)
rel_seq <- data.frame(psra.png.bac.aa.sediment@otu_table)
meta_full <- dplyr::left_join(data.frame(psra.png.bac.aa.sediment@sam_data), env.filtered, by = join_by(site == Name))

#Calculate distance matrix
rel_seq_distmat <- vegdist(rel_seq, method = "bray")
rel_seq_distmat <- as.matrix(rel_seq_distmat, labels = T)

#running NMDS in vegan (metaMDS)
#k=2
rel_seq_NMS <- metaMDS(rel_seq_distmat, distance = "bray", k = 2, maxit = 999, trymax = 100, wascores = TRUE)
rel_seq_NMS

#Goodness of fit/Shepards test
goodness(rel_seq_NMS) #produces a result of test statistics for goodness of fit for each point
stressplot(rel_seq_NMS)

#plot NMDS
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
envfit <- envfit(rel_seq_NMS ~ BO_cloudmean + BO21_salinitymean_ss + BO21_nitratemean_ss + BO21_tempmean_ss,
                 data=meta_full,
                 perm=999)
scores(envfit, "vectors")
plot(envfit)

svg("./Outputs/NMDS_w_env-aa-sediment-PNG-16s.svg")
plot(rel_seq_NMS, "sites")
orditorp(rel_seq_NMS, "sites")
plot(envfit)
dev.off()
```

# RDA Analysis
## Bacteria
### Model 1 
Need to attempt with only specific section of the organism; consider method to test significance with a categorical variable: hostSection to identify important ASVs.
```{r}
library(vegan)
library(psych)
library(adegenet)

ps.object <- psra.png.bac.aa

otu.data <- data.frame(otu_table(ps.object))
dim(otu.data)
sum(is.na(otu.data)) # Ensure no NAs. Impute NAs otherwise.
otu.taxa <- data.frame(tax_table(ps.object))
meta.data <- data.frame(sample_data(ps.object))

env.model = env.filtered
env.norm = as.data.frame(scale(env.model[,-1])) # Normalising environmental data
env.norm = data.frame(site = env.model[,1], env.norm)
env.data <- dplyr::left_join(meta.aa, env.norm, by = join_by(site))

rownames(env.data) <- env.data[,"sampleName"]
env.data <- env.data[,-c(1:10)]
str(env.data)

identical(rownames(otu.data), rownames(env.data)) # Ensuring rows are in order.

pairs.panels(env.data, scale=T) # Checking for correlations
svg("./Outputs/RDA/model_1_env_correlations-aa-PNG-16s.svg") #EDIT
pairs.panels(env.data, scale=T)
dev.off()

pred <- subset(env.data, select = colnames(env.data)) # Choosing desired variables
pairs.panels(pred, scale=T)
svg("./Outputs/RDA/model_1_env_correlations_selected-aa-PNG-16s.svg") #EDIT
pairs.panels(pred, scale=T)
dev.off()

# RDA analysis
bac.rda <- rda(otu.data ~ ., data=pred, scale=T) # tempmean aliased
bac.rda

# Estimate the variance explained by the RDA
RsquareAdj(bac.rda) #0.028
summary(eigenvals(bac.rda, model = "constrained"))
screeplot(bac.rda) 
svg("./Outputs/RDA/model_1_RDA_screeplot-aa-PNG-16s.svg")
screeplot(bac.rda) 
dev.off()
signif.full <- anova.cca(bac.rda, parallel=getOption("mc.cores")) 
signif.full#p=0.001

# Estimate the variance inflation factors of the variables - if this measure is over 5, it suggests there is still significant correlation among factors and the model is overfit.
vif.cca(bac.rda) # all less than 2, tempmean NA

# Plotting the RDA
plot(bac.rda, scaling=3)
plot(bac.rda, choices = c(1, 3), scaling=3)

# Identifying OTUs that vary significantly with the environmental variables
load.rda <- scores(bac.rda, choices=c(1:2), display="species")

# Examine the loadings for each RDA axis. These should follow an approximate normal distribution
hist(load.rda[,1], main="Loadings on RDA1") 
hist(load.rda[,2], main="Loadings on RDA2")

# Find loadings that are significantly associated with the bacteria
outliers <- function(x,z){
  lims <- mean(x) + c(-1, 1) * z * sd(x)          
  x[x < lims[1] | x > lims[2]]}

cand1 <- outliers(load.rda[,1],3) 
cand2 <- outliers(load.rda[,2],3) 

ncand <- length(cand1) + length(cand2)
ncand

cand1 <- cbind.data.frame(rep(1,times=length(cand1)), names(cand1), unname(cand1))
cand2 <- cbind.data.frame(rep(2,times=length(cand2)), names(cand2), unname(cand2))
colnames(cand1) <- colnames(cand2)  <- c("axis","otu","loading")

cand <- rbind(cand1, cand2)
cand$otu <- as.character(cand$otu)
cand <- cand[order(abs(cand$loading), decreasing = TRUE),]

write.table(cand, "./Outputs/RDA/model_1_RDA_OTUs-aa-PNG-16s.txt",  quote=F) #EDIT
cand


# Organise the data for generating a plot that shows the candidate otus and what environment it is associated with
head(pred)

foo <- matrix(nrow=(ncand), ncol=ncol(pred))  # n columns for n predictors change this depending on number of predictors
colnames(foo) <- colnames(pred)  ## this will also vary depending on your predictors

for (i in 1:length(cand$otu)) {
  nam <- cand[i,2]
  otu.gen <- otu.data[,nam]
  foo[i,] <- apply(pred,2,function(x) cor(x,otu.gen))
}

cand <- cbind.data.frame(cand,foo)  
head(cand)

length(cand$otu[duplicated(cand$otu)])

for (i in 1:length(cand$otu)) {
  bar <- cand[i,]
  cand[i,(ncol(pred)+4)] <- names(which.max(abs(bar[4:(ncol(pred)+3)]))) # gives the variable
  cand[i,(ncol(pred)+5)] <- max(abs(bar[4:(ncol(pred)+3)]))              # gives the correlation
}

colnames(cand)[(ncol(pred)+4)] <- "predictor"
colnames(cand)[(ncol(pred)+5)] <- "correlation"

table(cand$predictor) 

write.table(cand, "./Outputs/RDA/model_1_OTU_env_correlations-aa-PNG-16s.txt", quote = F) #EDIT
cand
unique(cand$predictor)

sel <- cand$otu
bac <- cand$predictor
colorScheme <- data.frame(
  Variable = unique(cand$predictor),
  Color = c('#00cccc', '#cc0066', '#ffcc66', '#6600cc') #EDIT. Should equal the number of unique predictors. Extra colors: 
)
for(i in 1: nrow(colorScheme)){
  bac[bac==colorScheme$Variable[i]] <- colorScheme$Color[i]
}


# color by predictor:
col.pred <- rownames(bac.rda$CCA$v) # pull the OTU names
asv.pred <- ifelse(col.pred %in% sel, otu.taxa[col.pred,"Phylum"], "") # Choose which taxonomic level to display

for (i in 1:length(sel)) {           # color code candidate OTUs
  foo <- match(sel[i],col.pred)
  col.pred[foo] <- bac[i]
}

col.pred[!(col.pred %in% bac)] <- '#f1eef6' # non-candidate OTUs
empty <- col.pred
empty[grep("#f1eef6",empty)] <- rgb(0,1,0, alpha=0) # transparent
empty.outline <- ifelse(empty=="#00FF0000","#00FF0000","gray32")
bg <- colorScheme$Color
sizeKey <- data.frame(
  Color = c(colorScheme$Color, "#f1eef6"),
  Size = c(rep(1.5, nrow(colorScheme)), 1)
)
size.pred <- recode(col.pred, !!!setNames(sizeKey$Size, sizeKey$Color))

# color the arrows
labels.pred <- colnames(pred)
for (i in 1:nrow(colorScheme)) {           # color code candidate OTUs
  labels.pred[labels.pred == colorScheme$Variable[i]] <- colorScheme$Color[i]
}
labels.pred[!(labels.pred %in% colorScheme$Color)] <- "#0868ac" # non-candidate OTUs

# color the samples
samNames <- rownames(otu.data)
samLocation <- data.frame(ps.object@sam_data)[samNames,"site"]
samColor <- paste0(COLOR_SITE[samLocation], "75")

# Plot axes 1 & 2 for the RDA
plot(bac.rda, type="n", scaling=3, #xlim=c(-0.5,2), ylim=c(-1,0.5),
     main = "Avicennea alba Bacterial Predictors of Environment")
points(bac.rda, display="species", pch=21, cex=size.pred, col="gray32", bg=col.pred, scaling=3)
text(bac.rda, scaling=3, display="species", cex=0.3, labels = asv.pred, pos=1, offset=0.4)
points(bac.rda, display="species", pch=21, cex=size.pred, col=empty.outline, bg=empty, scaling=3)
points(bac.rda, display="sites", pch=24, cex=1, col="gray32", bg="#0868ac", scaling=3)
text(bac.rda, scaling=3, display="sites", cex=0.3, pos=1, offset=0.4)
text(bac.rda, scaling=3, display="bp", col=labels.pred, cex=1)
legend("bottomright", legend=colorScheme$Variable, bty="n", col="gray32", pch=21, cex=1, pt.bg=bg)

svg("./Outputs/RDA/model_1_Env_RDA_Plot-aa-PNG-16s.svg")
plot(bac.rda, type="n", scaling=3, xlim=c(-2,2), ylim=c(-1,1),
     main = paste0("Avicennea alba\nBacterial Predictors of Environment\nModel 1 - R^2 = ", round(RsquareAdj(bac.rda)$adj.r.squared, digits = 2)))
points(bac.rda, display="species", pch=21, cex=size.pred, col="gray32", bg=col.pred, scaling=3)
text(bac.rda, scaling=3, display="species", cex=0.3, labels = asv.pred, pos=1, offset=0.4)
points(bac.rda, display="species", pch=21, cex=size.pred, col=empty.outline, bg=empty, scaling=3)
points(bac.rda, display="sites", pch=24, cex=1, col="#52525275", bg=samColor, scaling=3)
#text(bac.rda, scaling=3, display="sites", cex=0.3, pos=1, offset=0.4)
text(bac.rda, scaling=3, display="bp", col=labels.pred, cex=1)
legend("bottomright", legend=colorScheme$Variable, bty="n", col="gray32", pch=21, cex=1, pt.bg=bg, title = "Environmental Variable and Predictors")
legend("bottomleft", legend=sort(unique(meta.data$site)), bty="n", col="#52525275", pch=24, cex=1, pt.bg=paste0(COLOR_SITE[sort(unique(meta.data$site))], "75"), title = "Sample Site")
dev.off()


############# TRYING TO IMPROVE THE MODEL ############# (untouched)

env.raw = read.csv("./sdmpredictorsData/PNG_env_extract_all.csv", row.names = 1)
env.model = env.raw
env.norm = as.data.frame(scale(env.model[,-1])) # Normalising environmental data
env.norm = data.frame(site = env.model[,1], env.norm)
env.data <- dplyr::left_join(meta.aa, env.norm, by = join_by(site))
rownames(env.data) <- env.data[,"sampleName"]
env.data <- env.data[,-c(1:10)]
str(env.data)
pred <- subset(env.data, select = colnames(env.data)) # Choosing desired variables

fwd.sel <- ordiR2step(rda(otu.data ~ 1, data = pred), # lower model limit (simple!)
               scope = formula(bac.rda), # upper model limit (the "full" model)
               direction = "forward",
               R2scope = FALSE, # can't surpass the "full" model's R2
               pstep = 1000,
               trace = TRUE) # change to TRUE to see the selection process!

fwd.sel$call

# Write our new model
pred <- subset(env.data, select= c("BO21_salinitymean_ss", "BO_parmean")) # Choosing desired variables
bac.rda.signif <- rda(formula = otu.data ~ ., data = pred)
# check the adjusted R2 (corrected for the number of explanatory variables)
bac.rda.signif
RsquareAdj(bac.rda.signif)

# significance testing
anova.cca(bac.rda, step = 1000)
anova.cca(bac.rda.signif, step = 1000)

anova.cca(bac.rda, step = 1000, by = "term")
anova.cca(bac.rda.signif, step = 1000, by = "term")

anova.cca(bac.rda.signif, step = 1000, by = "axis")

# Estimate the variance explained by the RDA
RsquareAdj(bac.rda.signif)
summary(eigenvals(bac.rda.signif, model = "constrained"))
screeplot(bac.rda.signif) 
svg("./Outputs/Analysis/RDA/model_1_Bac_RDA_signif_screeplot.svg")
screeplot(bac.rda.signif) 
dev.off()
signif.full.signif <- anova.cca(bac.rda.signif, parallel=getOption("mc.cores")) 
signif.full.signif

# Estimate the variance inflation factors of the variables - if this measure is over 5, it suggests there is still significant correlation among factors and the model is overfit.
vif.cca(bac.rda.signif)

# Plotting the RDA
plot(bac.rda.signif, scaling=3)
plot(bac.rda.signif, choices = c(1, 3), scaling=3)

# Identifying Bacterial OTUs that vary significantly with the Sym OTUs
load.rda <- scores(bac.rda.signif, choices=c(1:2), display="species")

# Examine the loadings for each RDA axis. These should follow an approximate normal distribution
hist(load.rda[,1], main="Loadings on RDA1") 
hist(load.rda[,2], main="Loadings on RDA2")

# Find loadings that are significantly associated with the symbiodiniaceae
cand1 <- outliers(load.rda[,1],3) 
cand2 <- outliers(load.rda[,2],3) 

ncand <- length(cand1) + length(cand2)
ncand

cand1 <- cbind.data.frame(rep(1,times=length(cand1)), names(cand1), unname(cand1))
cand2 <- cbind.data.frame(rep(2,times=length(cand2)), names(cand2), unname(cand2))
colnames(cand1) <- colnames(cand2)  <- c("axis","otu","loading")

cand <- rbind(cand1, cand2)
cand$otu <- as.character(cand$otu)
cand <- cand[order(abs(cand$loading), decreasing = TRUE),]

write.table(cand, "./Outputs/Analysis/RDA/model_1_Bac_Pl_RDA_OTUs_signif.txt",  quote=F)
cand


# Organise the data for generating a plot that shows the candidate otus and what environment it is associated with
head(pred)

foo <- matrix(nrow=(ncand), ncol=ncol(pred))  # n columns for n predictors change this depending on number of predictors
colnames(foo) <- colnames(pred)  ## this will also vary depending on your predictors

for (i in 1:length(cand$otu)) {
  nam <- cand[i,2]
  otu.gen <- otu.data[,nam]
  foo[i,] <- apply(pred,2,function(x) cor(x,otu.gen))
}

cand <- cbind.data.frame(cand,foo)  
head(cand)

length(cand$otu[duplicated(cand$otu)])

for (i in 1:length(cand$otu)) {
  bar <- cand[i,]
  cand[i,(ncol(pred)+4)] <- names(which.max(abs(bar[4:(ncol(pred)+3)]))) # gives the variable
  cand[i,(ncol(pred)+5)] <- max(abs(bar[4:(ncol(pred)+3)]))              # gives the correlation
}

colnames(cand)[(ncol(pred)+4)] <- "predictor"
colnames(cand)[(ncol(pred)+5)] <- "correlation"

table(cand$predictor) 

write.table(cand, "./Outputs/Analysis/RDA/model_1_Bac_Pl_OTU_env_correlations_signif.txt", quote = F)
cand
unique(cand$predictor)

sel <- cand$otu
sym <- cand$predictor
colorScheme <- data.frame(
  Variable = unique(cand$predictor),
  Color = c('#00cccc', '#cc0066')
)
for(i in 1: nrow(colorScheme)){
  sym[sym==colorScheme$Variable[i]] <- colorScheme$Color[i]
}


# color by predictor:
col.pred <- rownames(bac.rda.signif$CCA$v) # pull the OTU names
asv.pred <- ifelse(col.pred %in% sel, otu.taxa[col.pred,"Phylum"], "")

for (i in 1:length(sel)) {           # color code candidate OTUs
  foo <- match(sel[i],col.pred)
  col.pred[foo] <- sym[i]
}

col.pred[!(col.pred %in% sym)] <- '#f1eef6' # non-candidate OTUs
empty <- col.pred
empty[grep("#f1eef6",empty)] <- rgb(0,1,0, alpha=0) # transparent
empty.outline <- ifelse(empty=="#00FF0000","#00FF0000","gray32")
bg <- colorScheme$Color
sizeKey <- data.frame(
  Color = c(colorScheme$Color, "#f1eef6"),
  Size = c(rep(1.5, nrow(colorScheme)), 1)
)
size.pred <- recode(col.pred, !!!setNames(sizeKey$Size, sizeKey$Color))

# color the arrows
labels.pred <- colnames(pred)
for (i in 1:nrow(colorScheme)) {           # color code candidate OTUs
  labels.pred[labels.pred == colorScheme$Variable[i]] <- colorScheme$Color[i]
}
labels.pred[!(labels.pred %in% colorScheme$Color)] <- "#0868ac" # non-candidate OTUs

# color the samples
samNames <- rownames(otu.data)
samLocation <- data.frame(ps.object@sam_data)[samNames,"Site"]
samColor <- paste0(col_site[samLocation], "75")

# Plot axes 1 & 2 for the RDA
plot(bac.rda.signif, type="n", scaling=3, #xlim=c(-0.5,2), ylim=c(-1,0.5),
     main = "Porites Lutea Bacterial Predictors of Environment (Significant)")
points(bac.rda.signif, display="species", pch=21, cex=size.pred, col="gray32", bg=col.pred, scaling=3)
text(bac.rda.signif, scaling=3, display="species", cex=0.3, labels = asv.pred, pos=1, offset=0.4)
points(bac.rda.signif, display="species", pch=21, cex=size.pred, col=empty.outline, bg=empty, scaling=3)
points(bac.rda.signif, display="sites", pch=24, cex=1, col="gray32", bg="#0868ac", scaling=3)
text(bac.rda.signif, scaling=3, display="sites", cex=0.3, pos=1, offset=0.4)
text(bac.rda.signif, scaling=3, display="bp", col=labels.pred, cex=1)
legend("bottomright", legend=colorScheme$Variable, bty="n", col="gray32", pch=21, cex=1, pt.bg=bg)

svg("./Outputs/Analysis/RDA/model_1_Bac_Pl_Env_RDA_signif_Plot.svg")
plot(bac.rda.signif, type="n", scaling=3, xlim=c(-1,1), ylim=c(-0.5,0.5),
     main = paste0("Porites Lutea\nBacterial Predictors of Environment\nSignificant Model 1 - R^2 = ", round(RsquareAdj(bac.rda.signif)$adj.r.squared, digits = 2)))
points(bac.rda.signif, display="species", pch=21, cex=size.pred, col="gray32", bg=col.pred, scaling=3)
text(bac.rda.signif, scaling=3, display="species", cex=0.3, labels = asv.pred, pos=1, offset=0.4)
points(bac.rda.signif, display="species", pch=21, cex=size.pred, col=empty.outline, bg=empty, scaling=3)
points(bac.rda.signif, display="sites", pch=24, cex=1, col="#52525275", bg=samColor, scaling=3)
#text(bac.rda.signif, scaling=3, display="sites", cex=0.3, pos=1, offset=0.4)
text(bac.rda.signif, scaling=3, display="bp", col=labels.pred, cex=1, labels = "")
legend("bottomright", legend=colorScheme$Variable, bty="n", col="gray32", pch=21, cex=1, pt.bg=bg, title = "Environmental Variable and Predictors")
legend("bottomleft", legend=names(col_site), bty="n", col="#52525275", pch=24, cex=1, pt.bg=paste0(col_site,"75"), title = "Sample Site")
dev.off()

```


# FAPROTAX
```{r}
# To run if needed

pasteTaxa <- function(taxa_table){
  return(paste(taxa_table, collapse = "; "))
}

# For full taxonomy only table
ps.png.bac
taxa.table.full <- data.frame(tax_table(ps.png.bac))
taxa.table.full.simple <- data.frame(ASV_ID = rownames(taxa.table.full), taxonomy = base::apply(taxa.table.full, 1, FUN = pasteTaxa))
write_tsv(taxa.table.full.simple, "./Outputs/FAPROTAX/input_faprotax-taxa_full-PNG-16s.tsv") # Edit output file: add # before ASV_ID.
# Code run on terminal from FAPROTAX download folder
# python collapse_table.py -i ../Outputs/FAPROTAX/input_faprotax-taxa_full-PNG-16s.tsv -o ../Outputs/FAPROTAX/functional_table-taxa_full-PNG-16s.tsv -g FAPROTAX.txt -c "#" -d "taxonomy" --omit_columns 0 --column_names_are_in last_comment_line -r ../Outputs/FAPROTAX/report-taxa_full-PNG-16s.txt -n columns_after_collapsing -v
```

```{r}
# To run if needed

# For all samples
ps.png.bac
sample.table.full <- data.frame(sample_data(ps.png.bac))
otu.table.full <- data.frame(otu_table(ps.png.bac, taxa_are_rows = FALSE))
otu.table.full <- data.frame(t(otu.table.full))
taxa.table.full <- data.frame(tax_table(ps.png.bac))
taxa.table.full.simple <- data.frame(ASV_ID = rownames(taxa.table.full), taxonomy = base::apply(taxa.table.full, 1, FUN = pasteTaxa))
faprotax.table.full <- data.frame(otu.table.full[rownames(taxa.table.full.simple),], taxonomy = taxa.table.full.simple$taxonomy)
write_tsv(faprotax.table.full, "./Outputs/FAPROTAX/input_faprotax-PNG-16s.tsv")
# python collapse_table.py -i ../Outputs/FAPROTAX/input_faprotax-PNG-16s.tsv -o ../Outputs/FAPROTAX/functional_table-PNG-16s.tsv -g FAPROTAX.txt -c "#" -d "taxonomy" -r ../Outputs/FAPROTAX/report-PNG-16s.txt -n columns_after_collapsing -v
```

```{r}
function_table.png.16s <- read_tsv("./Outputs/FAPROTAX/functional_table-PNG-16s.tsv")
function_table.png.16s <- function_table.png.16s %>% 
  column_to_rownames(var = "group") %>% 
  t() %>% 
  data.frame()
rownames(function_table.png.16s) <- sample.table.full$sampleName

function_table.png.16s.long <- pivot_longer(
  data.frame(sampleName = rownames(function_table.png.16s), function_table.png.16s, sample.table.full[rownames(function_table.png.16s),]),
  cols = colnames(function_table.png.16s),
  names_to = "functionalGroup"
)
function_table.png.16s.long$sampleName.1 <- NULL


ggplot(function_table.png.16s.long %>% 
         mutate(sampleName = fct_reorder(sampleName, dplyr::desc(3 * as.numeric(hostTaxa == "Coral") + 
                                                                 2 * as.numeric(hostTaxa == "Seagrass") +
                                                                 1 * as.numeric(hostSection == "Mangrove")
                                                                 )))) + 
  geom_tile(aes(x = sampleName, y = functionalGroup, fill = value)) + 
  theme(axis.text.x = element_text(size = 0.75, angle = 30, vjust = .8, hjust = 0.8))


# For Avicinea alba
function_table.aa.png.16s.long <- base::subset(function_table.png.16s.long, host == "Avicinea alba")
unique(function_table.aa.png.16s.long$hostSection)
ggplot(function_table.png.16s.long %>% 
         mutate(sampleName = fct_reorder(sampleName, dplyr::desc(4 * as.numeric(hostSection == "Fruit") + 
                                                                 3 * as.numeric(hostSection == "Leaf") +
                                                                 2 * as.numeric(hostSection == "Pneumatophore") +
                                                                 1 * as.numeric(hostSection == "Sediment")
                                                                 )))) + 
  geom_tile(aes(x = sampleName, y = functionalGroup, fill = value)) + 
  theme(axis.text.x = element_text(size = 0.75, angle = 30, vjust = .8, hjust = 0.8))

ggplot(function_table.aa.png.16s.long) + 
  geom_boxplot(aes(y = value, x = functionalGroup, color = hostSection)) +
  theme(axis.text.x = element_text(size = 5, angle = -30, vjust = 1, hjust = 0))

functional_group_significance <- data.frame()
for(i in unique(function_table.aa.png.16s.long$functionalGroup)){
  df <- subset(function_table.aa.png.16s.long, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$hostSection, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
}
functional_group_significance

significant_functional_groups <- functional_group_significance[which(functional_group_significance[,2] < 0.05),1]
ggplot(subset(function_table.aa.png.16s.long, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value, y = hostSection, color = hostSection)) +
  facet_wrap(vars(functionalGroup), scales = "free")
```

```{r}
core_taxa <- core_microbiome$ASVID
DA_taxa <- unique(c(
  read.csv("./Outputs/DAtaxalist_hostSection-mangrove-PNG-16S.csv")$x,
  read.csv("./Outputs/DAtaxalist_host-mangrove-PNG-16S.csv")$x,
  read.csv("./Outputs/DAtaxalist_site-mangrove-PNG-16S.csv")$x))
```

```{r}
# All mangrove
psra.png.bac.mangrove <- prune_taxa(taxa_sums(psra.png.bac.mangrove) > 0, psra.png.bac.mangrove)
sample.table.full <- data.frame(sample_data(psra.png.bac.mangrove))
otu.table.full <- data.frame(otu_table(psra.png.bac.mangrove, taxa_are_rows = FALSE))
otu.table.full <- data.frame(t(otu.table.full))
taxa.table.full <- data.frame(tax_table(psra.png.bac.mangrove))
taxa.table.full.simple <- data.frame(ASV_ID = rownames(taxa.table.full), taxonomy = base::apply(taxa.table.full, 1, FUN = pasteTaxa))
faprotax.table.full <- data.frame(otu.table.full[rownames(taxa.table.full.simple),], taxonomy = taxa.table.full.simple$taxonomy)
write_tsv(faprotax.table.full, "./Outputs/FAPROTAX/input_faprotax-mangrove-PNG-16s.tsv")
# python collapse_table.py -i ../Outputs/FAPROTAX/input_faprotax-mangrove-PNG-16s.tsv -o ../Outputs/FAPROTAX/functional_table-mangrove-PNG-16s.tsv -g FAPROTAX.txt -c "#" -d "taxonomy" -r ../Outputs/FAPROTAX/report-mangrove-PNG-16s.txt -n columns_after_collapsing -v

function_table <- read_tsv("./Outputs/FAPROTAX/functional_table-mangrove-PNG-16s.tsv")
function_table <- function_table %>% 
  column_to_rownames(var = "group") %>% 
  t() %>% 
  data.frame()
rownames(function_table) <- sample.table.full$sampleName

function_table <- pivot_longer(
  data.frame(sampleName = rownames(function_table), function_table, sample.table.full[rownames(function_table),]),
  cols = colnames(function_table),
  names_to = "functionalGroup"
)
function_table$sampleName.1 <- NULL

ggplot(function_table) + 
  geom_boxplot(aes(x = value, y = functionalGroup, color = hostSection)) #+ facet_wrap(vars(functionalGroup), scales = "free")

functional_group_significance <- data.frame()
for(i in unique(function_table$functionalGroup)){
  df <- subset(function_table, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$hostSection, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
}
functional_group_significance

function_table$value_scaled <- ave(function_table$value, function_table$functionalGroup, FUN=scale)
significant_functional_groups <- functional_group_significance[which(as.numeric(functional_group_significance[,2]) < 0.05),1]

ggplot(subset(function_table, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value_scaled, y = functionalGroup, color = hostSection)) + scale_color_manual(values = COLOR_HOST_SECTION) + labs(x = "Z score of relative abundance within functional group") #+ facet_wrap(vars(functionalGroup), scales = "free")
ggsave(filename = "./Outputs/faprotaxAnalysis_hostSection-mangrove-PNG-16S.svg", dpi = 300, width = 12, height = 8)

functional_group_significance <- data.frame()
for(i in unique(function_table$functionalGroup)){
  df <- subset(function_table, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$host, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
}
functional_group_significance
function_table$value_scaled <- ave(function_table$value, function_table$functionalGroup, FUN=scale)
significant_functional_groups <- functional_group_significance[which(as.numeric(functional_group_significance[,2]) < 0.05),1]

ggplot(subset(function_table, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value_scaled, y = functionalGroup, color = host)) + scale_color_manual(values = COLOR_HOST) + labs(x = "Z score of relative abundance within functional group") #+ facet_wrap(vars(functionalGroup), scales = "free")
ggsave(filename = "./Outputs/faprotaxAnalysis_host-mangrove-PNG-16S.svg", dpi = 300, width = 12, height = 8)

functional_group_significance <- data.frame()
for(i in unique(function_table$functionalGroup)){
  df <- subset(function_table, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$site, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
}
functional_group_significance
function_table$value_scaled <- ave(function_table$value, function_table$functionalGroup, FUN=scale)
significant_functional_groups <- functional_group_significance[which(as.numeric(functional_group_significance[,2]) < 0.05),1]
ggplot(subset(function_table, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value_scaled, y = functionalGroup, color = site)) + scale_color_manual(values = COLOR_SITE) + labs(x = "Z score of relative abundance within functional group") #+ facet_wrap(vars(functionalGroup), scales = "free")
ggsave(filename = "./Outputs/faprotaxAnalysis_site-mangrove-PNG-16S.svg", dpi = 300, width = 12, height = 8)
```

```{r}
# Core mangrove
psra.png.bac.mangrove
psra.png.bac.mangrove.core <- prune_taxa(taxa_names(psra.png.bac.mangrove) %in% core_taxa, psra.png.bac.mangrove)
sample.table.full <- data.frame(sample_data(psra.png.bac.mangrove.core))
otu.table.full <- data.frame(otu_table(psra.png.bac.mangrove.core, taxa_are_rows = FALSE))
otu.table.full <- data.frame(t(otu.table.full))
taxa.table.full <- data.frame(tax_table(psra.png.bac.mangrove.core))
taxa.table.full.simple <- data.frame(ASV_ID = rownames(taxa.table.full), taxonomy = base::apply(taxa.table.full, 1, FUN = pasteTaxa))
faprotax.table.full <- data.frame(otu.table.full[rownames(taxa.table.full.simple),], taxonomy = taxa.table.full.simple$taxonomy)
write_tsv(faprotax.table.full, "./Outputs/FAPROTAX/input_faprotax-core-mangrove-PNG-16s.tsv")
# python collapse_table.py -i ../Outputs/FAPROTAX/input_faprotax-core-mangrove-PNG-16s.tsv -o ../Outputs/FAPROTAX/functional_table-core-mangrove-PNG-16s.tsv -g FAPROTAX.txt -c "#" -d "taxonomy" -r ../Outputs/FAPROTAX/report-core-mangrove-PNG-16s.txt -n columns_after_collapsing -v

function_table <- read_tsv("./Outputs/FAPROTAX/functional_table-core-mangrove-PNG-16s.tsv")
function_table <- function_table %>% 
  column_to_rownames(var = "group") %>% 
  t() %>% 
  data.frame()
rownames(function_table) <- sample.table.full$sampleName

function_table <- pivot_longer(
  data.frame(sampleName = rownames(function_table), function_table, sample.table.full[rownames(function_table),]),
  cols = colnames(function_table),
  names_to = "functionalGroup"
)
function_table$sampleName.1 <- NULL

ggplot(function_table) + 
  geom_boxplot(aes(x = value, y = functionalGroup, color = hostSection)) #+ facet_wrap(vars(functionalGroup), scales = "free")

functional_group_significance <- data.frame()
for(i in unique(function_table$functionalGroup)){
  df <- subset(function_table, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$hostSection, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
}
functional_group_significance
function_table$value_scaled <- ave(function_table$value, function_table$functionalGroup, FUN=scale)
significant_functional_groups <- functional_group_significance[which(as.numeric(functional_group_significance[,2]) < 0.05),1]

ggplot(subset(function_table, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value_scaled, y = functionalGroup, color = hostSection)) + scale_color_manual(values = COLOR_HOST_SECTION) + labs(x = "Z score of relative abundance within functional group") #+ facet_wrap(vars(functionalGroup), scales = "free")
ggsave(filename = "./Outputs/faprotaxAnalysis_hostSection-core-mangrove-PNG-16S.svg", dpi = 300, width = 12, height = 8)

functional_group_significance <- data.frame()
for(i in unique(function_table$functionalGroup)){
  df <- subset(function_table, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$host, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
}
functional_group_significance
function_table$value_scaled <- ave(function_table$value, function_table$functionalGroup, FUN=scale)
significant_functional_groups <- functional_group_significance[which(as.numeric(functional_group_significance[,2]) < 0.05),1]

ggplot(subset(function_table, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value_scaled, y = functionalGroup, color = host)) + scale_color_manual(values = COLOR_HOST) + labs(x = "Z score of relative abundance within functional group") #+ facet_wrap(vars(functionalGroup), scales = "free")
ggsave(filename = "./Outputs/faprotaxAnalysis_host-core-mangrove-PNG-16S.svg", dpi = 300, width = 12, height = 8)

functional_group_significance <- data.frame()
for(i in unique(function_table$functionalGroup)){
  df <- subset(function_table, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$site, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
}
functional_group_significance
function_table$value_scaled <- ave(function_table$value, function_table$functionalGroup, FUN=scale)
significant_functional_groups <- functional_group_significance[which(as.numeric(functional_group_significance[,2]) < 0.05),1]

ggplot(subset(function_table, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value_scaled, y = functionalGroup, color = site)) + scale_color_manual(values = COLOR_SITE) + labs(x = "Z score of relative abundance within functional group") #+ facet_wrap(vars(functionalGroup), scales = "free")
ggsave(filename = "./Outputs/faprotaxAnalysis_site-core-mangrove-PNG-16S.svg", dpi = 300, width = 12, height = 8)
```

```{r}
# DA mangrove
psra.png.bac.mangrove
psra.png.bac.mangrove.DA <- prune_taxa(taxa_names(psra.png.bac.mangrove) %in% DA_taxa, psra.png.bac.mangrove)
sample.table.full <- data.frame(sample_data(psra.png.bac.mangrove.DA))
otu.table.full <- data.frame(otu_table(psra.png.bac.mangrove.DA, taxa_are_rows = FALSE))
otu.table.full <- data.frame(t(otu.table.full))
taxa.table.full <- data.frame(tax_table(psra.png.bac.mangrove.DA))
taxa.table.full.simple <- data.frame(ASV_ID = rownames(taxa.table.full), taxonomy = base::apply(taxa.table.full, 1, FUN = pasteTaxa))
faprotax.table.full <- data.frame(otu.table.full[rownames(taxa.table.full.simple),], taxonomy = taxa.table.full.simple$taxonomy)
write_tsv(faprotax.table.full, "./Outputs/FAPROTAX/input_faprotax-DA-mangrove-PNG-16s.tsv")
# python collapse_table.py -i ../Outputs/FAPROTAX/input_faprotax-DA-mangrove-PNG-16s.tsv -o ../Outputs/FAPROTAX/functional_table-DA-mangrove-PNG-16s.tsv -g FAPROTAX.txt -c "#" -d "taxonomy" -r ../Outputs/FAPROTAX/report-DA-mangrove-PNG-16s.txt -n columns_after_collapsing -v

function_table <- read_tsv("./Outputs/FAPROTAX/functional_table-DA-mangrove-PNG-16s.tsv")
function_table <- function_table %>% 
  column_to_rownames(var = "group") %>% 
  t() %>% 
  data.frame()
rownames(function_table) <- sample.table.full$sampleName

function_table <- pivot_longer(
  data.frame(sampleName = rownames(function_table), function_table, sample.table.full[rownames(function_table),]),
  cols = colnames(function_table),
  names_to = "functionalGroup"
)
function_table$sampleName.1 <- NULL

ggplot(function_table) + 
  geom_boxplot(aes(x = value, y = functionalGroup, color = hostSection)) #+ facet_wrap(vars(functionalGroup), scales = "free")

functional_group_significance <- data.frame()
for(i in unique(function_table$functionalGroup)){
  df <- subset(function_table, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$hostSection, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
  if (min_p_value < 0.05){
    print(i)
    print(pairwise.wilcox.test(x = df$value, g = df$hostSection, p.adjust.method = "bonferroni"))
  }
}
functional_group_significance
function_table$value_scaled <- ave(function_table$value, function_table$functionalGroup, FUN=scale)
significant_functional_groups <- functional_group_significance[which(as.numeric(functional_group_significance[,2]) < 0.05),1]

ggplot(subset(function_table, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value_scaled, y = functionalGroup, color = hostSection)) + scale_color_manual(values = COLOR_HOST_SECTION) + labs(x = "Z score of relative abundance within functional group") #+ facet_wrap(vars(functionalGroup), scales = "free")
ggsave(filename = "./Outputs/faprotaxAnalysis_hostSection-DA-mangrove-PNG-16S.svg", dpi = 300, width = 12, height = 8)

functional_group_significance <- data.frame()
for(i in unique(function_table$functionalGroup)){
  df <- subset(function_table, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$host, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
}
functional_group_significance
function_table$value_scaled <- ave(function_table$value, function_table$functionalGroup, FUN=scale)
significant_functional_groups <- functional_group_significance[which(as.numeric(functional_group_significance[,2]) < 0.05),1]

ggplot(subset(function_table, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value_scaled, y = functionalGroup, color = host)) + scale_color_manual(values = COLOR_HOST) + labs(x = "Z score of relative abundance within functional group") #+ facet_wrap(vars(functionalGroup), scales = "free")
ggsave(filename = "./Outputs/faprotaxAnalysis_host-DA-mangrove-PNG-16S.svg", dpi = 300, width = 12, height = 8)


functional_group_significance <- data.frame()
for(i in unique(function_table$functionalGroup)){
  df <- subset(function_table, functionalGroup == i)
  min_p_value <- min(pairwise.wilcox.test(x = df$value, g = df$site, p.adjust.method = "bonferroni")$p.value, na.rm = TRUE)
  functional_group_significance <- rbind(functional_group_significance, c(i, min_p_value))
}
functional_group_significance
function_table$value_scaled <- ave(function_table$value, function_table$functionalGroup, FUN=scale)
significant_functional_groups <- functional_group_significance[which(as.numeric(functional_group_significance[,2]) < 0.05),1]

ggplot(subset(function_table, functionalGroup %in% significant_functional_groups)) + 
  geom_boxplot(aes(x = value_scaled, y = functionalGroup, color = site)) + scale_color_manual(values = COLOR_SITE) + labs(x = "Z score of relative abundance within functional group") #+ facet_wrap(vars(functionalGroup), scales = "free")
ggsave(filename = "./Outputs/faprotaxAnalysis_site-DA-mangrove-PNG-16S.svg", dpi = 300, width = 12, height = 8)
```




# Supervised Learning
```{r}

```

# Random Forest Model ~ host section
```{r}
library(magrittr)
library(grid)
library(reshape2)
library(randomForest)

source("../git_repos/MicrobeMiseq-master/R/miseqR.R")

ps.png.bac.mangrove
psra.png.bac.mangrove
```

```{r}
# Using relative abundance
# Select our four sections; Fruit, Leaf, Pneumatophore, Sediment
psra.png.bac.mangrove %>%
  subset_samples(hostSection %in% c("Fruit", "Leaf", "Pneumatophore", "Sediment")) -> sections.scale   
```

```{r}
# Glom at phylum level, prune out taxa < 2%, and format
sections.scale %>%
  taxglom_and_melt(taxrank = "Phylum", prune = 0.02) -> sections.long

sections.long <- sections.long %>% 
  mutate(
    metadata_independent_name = paste(str_extract(sampleName, "[[:digit:]]+"), sep = "-")
  )

# Stacked barplot 
ggplot(sections.long, aes(x = metadata_independent_name, y = Abundance, fill = Phylum)) + 
  facet_grid(hostSection~site+host) + 
  geom_bar(stat = "identity") +
  geom_bar(
    stat = "identity", 
    position = "fill", 
    colour = "black", 
    show.legend = FALSE
  ) + 
  scale_fill_manual(values = COLOR_21) +
  guides(fill = guide_legend(reverse = TRUE, keywidth = 1, keyheight = 1)) +
  xlab("") +
  ylab("Relative Abundance (Phyla > 2%) \n") +
  ggtitle("Phylum composition of bacteria in PNG different host sections") 
```

```{r}
# NMDS scores using bray-curtis distance 
set.seed(1)
bray.nmds <- ordinate(physeq = sections.scale, method = "NMDS", distance = "bray")

# Plot NMDS and color samples by sites
plot_ordination(
  physeq = sections.scale, 
  ordination = bray.nmds, 
  color = "hostSection"
) + 
  geom_point(size = 3) + 
  ggtitle("Bacterial communities from different host sections")
```

```{r}
adonis.site <- phyloseq_to_adonis(
  physeq = sections.scale, 
  dist = "bray", 
  formula = "hostSection"
)
```

```{r}
# How many OTUs do we currently have? 
ntaxa(sections.scale)
```

```{r}
# Set prunescale 
prunescale = 0.0001

# Prune out rare OTUs by mean relative abundance set by prunescale
tax.mean <- taxa_sums(sections.scale)/nsamples(sections.scale)
sections.prune <- prune_taxa(tax.mean > prunescale*1, sections.scale)

sections.prune
```

```{r}
# Make a dataframe of training data with OTUs as column and samples as rows
predictors <- otu_table(sections.prune)
dim(predictors)
```

```{r}
# Make one column for our outcome/response variable 
response <- as.factor(sample_data(sections.prune)$hostSection)

# Combine them into 1 data frame
rf.data <- data.frame(response, predictors)
```

```{r}
set.seed(2)
mangrove_section.classify <- randomForest(response~., data = rf.data, ntree = 100)
print(mangrove_section.classify)
```

```{r}
# What variables are stored in the output?
names(mangrove_section.classify)
```

```{r}
# Make a data frame with predictor names and their importance
imp <- importance(mangrove_section.classify)
imp <- data.frame(predictors = rownames(imp), imp)

# Order the predictor levels by importance
imp.sort <- arrange(imp, desc(MeanDecreaseGini))
imp.sort$predictors <- factor(imp.sort$predictors, levels = imp.sort$predictors)

# Select the top 10 predictors
imp.20 <- imp.sort[1:20, ]

# ggplot
ggplot(imp.20, aes(x = predictors, y = MeanDecreaseGini)) +
  geom_bar(stat = "identity", fill = "indianred") +
  coord_flip() +
  ggtitle("Most important ASVs for classifying mangrove samples into host section")
```

```{r}
otunames <- imp.20$predictors
r <- rownames(tax_table(sections.scale)) %in% otunames
data.frame(tax_table(sections.scale)[r, ])

imp.20 <- data.frame(imp.20, data.frame(tax_table(sections.scale)[r, ])[imp.20$predictors,])

ggplot(imp.20, aes(x = predictors, y = MeanDecreaseGini)) +
  geom_bar(stat = "identity", fill = "indianred") +
  geom_text(aes(label = Phylum)) +
  coord_flip() +
  ggtitle("Most important ASVs for classifying mangrove samples into host section")
```

# Random Forest Model ~ sectionSource
```{r}
library(magrittr)
library(grid)
library(reshape2)
library(randomForest)

source("../git_repos/MicrobeMiseq-master/R/miseqR.R")

ps.png.bac.mangrove.DA
psra.png.bac.mangrove.DA
```

```{r}
# Using relative abundance
# Select our four sections; Fruit, Leaf, Pneumatophore, Sediment
psra.png.bac.mangrove.DA %>%
  subset_samples(sectionSource %in% c("Host Tissue", "Sediment")) -> hosts.scale
```

```{r}
# Glom at phylum level, prune out taxa < 2%, and format
hosts.scale %>%
  taxglom_and_melt(taxrank = "Phylum", prune = 0.02) -> hosts.long

hosts.long <- hosts.long %>% 
  mutate(
    metadata_independent_name = paste(str_extract(sampleName, "[[:digit:]]+"), sep = "-")
  )

# Stacked barplot 
ggplot(hosts.long, aes(x = metadata_independent_name, y = Abundance, fill = Phylum)) + 
  facet_grid(hostSection~site+host) + 
  geom_bar(stat = "identity") +
  geom_bar(
    stat = "identity", 
    position = "fill", 
    colour = "black", 
    show.legend = FALSE
  ) + 
  scale_fill_manual(values = COLOR_21) +
  guides(fill = guide_legend(reverse = TRUE, keywidth = 1, keyheight = 1)) +
  xlab("") +
  ylab("Relative Abundance (Phyla > 2%) \n") +
  ggtitle("Phylum composition of bacteria in PNG host tissue and sediment") 
```

```{r}
# NMDS scores using bray-curtis distance 
set.seed(1)
bray.nmds <- ordinate(physeq = hosts.scale, method = "NMDS", distance = "bray")

# Plot NMDS and color samples by sites
plot_ordination(
  physeq = hosts.scale, 
  ordination = bray.nmds, 
  color = "host"
) + 
  geom_point(size = 3) + 
  ggtitle("Bacterial communities from host tissue and sediment")
```

```{r}
adonis.site <- phyloseq_to_adonis(
  physeq = hosts.scale, 
  dist = "bray", 
  formula = "sectionSource"
)

```

```{r}
# How many OTUs do we currently have? 
ntaxa(hosts.scale)
```

```{r}
# Set prunescale 
prunescale = 0.0001

# Prune out rare OTUs by mean relative abundance set by prunescale
tax.mean <- taxa_sums(hosts.scale)/nsamples(hosts.scale)
hosts.prune <- prune_taxa(tax.mean > prunescale*1, hosts.scale)

hosts.prune
```

```{r}
# Make a dataframe of training data with OTUs as column and samples as rows
predictors <- otu_table(hosts.prune)
dim(predictors)
```

```{r}
# Make one column for our outcome/response variable 
response <- as.factor(sample_data(hosts.prune)$host)

# Combine them into 1 data frame
rf.data <- data.frame(response, predictors)
```

```{r}
set.seed(2)
mangrove_host.classify <- randomForest(response~., data = rf.data, ntree = 100)
print(mangrove_host.classify)
```

```{r}
# What variables are stored in the output?
names(mangrove_host.classify)
```

```{r}
# Make a data frame with predictor names and their importance
imp <- importance(mangrove_host.classify)
imp <- data.frame(predictors = rownames(imp), imp)

# Order the predictor levels by importance
imp.sort <- arrange(imp, desc(MeanDecreaseGini))
imp.sort$predictors <- factor(imp.sort$predictors, levels = imp.sort$predictors)

# Select the top 10 predictors
imp.20 <- imp.sort[1:20, ]

# ggplot
ggplot(imp.20, aes(x = predictors, y = MeanDecreaseGini)) +
  geom_bar(stat = "identity", fill = "indianred") +
  coord_flip() +
  ggtitle("Most important ASVs for classifying mangrove samples into mangrove or sediment")
```

```{r}
otunames <- imp.20$predictors
r <- rownames(tax_table(hosts.scale)) %in% otunames
data.frame(tax_table(hosts.scale)[r, ])

imp.20 <- data.frame(imp.20, data.frame(tax_table(hosts.scale)[r, ])[imp.20$predictors,])

ggplot(imp.20, aes(x = predictors, y = MeanDecreaseGini)) +
  geom_bar(stat = "identity", fill = "indianred") +
  geom_text(aes(label = Phylum)) +
  coord_flip() +
  ggtitle("Most important ASVs for classifying mangrove samples into mangrove or sediment")
```









